---
title: "Chapter3-Dv-metabarcoding"
output:
  pdf_document: default
  html_document: default
---

```{r knitr setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r environment setup, include=F}
## R(v3.5.1 - Win)
#.libPaths(c("C:\\Users\\501596\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win", .libPaths("C:\\Users\\501596\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win")))

## R(v3.5.1 - OSX)
.libPaths(c("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx", .libPaths("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx")))

## R(v3.5.1 - Ubuntu)
#.libPaths(c("/home/mb/Dropbox/PhD Hull/PhD docs/Thesis/R_stats/rlib3.5_unix", .libPaths("/home/mb/Dropbox/PhD Hull/PhD docs/Thesis/R_stats/rlib3.5_unix")))
```

```{r directory setup, include=FALSE}
dir()
rm(list=ls())
ls()
```

## Setting working directory and packages

The analysis was done in R (R Core Team, 2018) using the environment and packages as specified below, and in `log_session_packages.txt`.

```{r packages loading, include=F}
pack.list = c("AED","aod","ape","bipartite","devtools","dplyr","EcoSimR","ggplot2","ggsignif","gplots","grid","gridExtra","iNEXT","lattice","lme4","lmtest","MASS","mgcv","msa","network","phyloseq","plyr","reshape2","spaa","stringi","UpSetR","vegan","zoo")
new.packages = pack.list[!(pack.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages, repos="https://cran.rstudio.com")
lapply(pack.list, require, character.only=T)
```

```{r session}
sessionInfo()
```

The data used in this analysis were generated across 2 separate sequencing runs; one containing gut contentns and eDNA samples collected in May 2017, the second with samples collected in October 2017. Despite the date of the sequencing, all data frames and variable will be named after the month of collection for practicality.

```{r input_files}
infile_may17 = read.csv("data/CO1Aug17-merge-forwonly-nonchimera-c97-cov2_id97-by-taxonomy-readcounts.blast.csv", header=T)
infile_oct17 = read.csv("data/CO1DvAug18-merged-forwonly_nonchimera-c97-cov2-id97_by-taxonomy-readcounts.csv", header=T)
```

Removing the taxonomy column from each input files and saving them into separate files.

```{r removing_taxonomy}
## Custom function to format the taxonomy column

format_taxonomy <- function(input_data){
  for (t in as.data.frame(input_data)){
    n = gsub("\\['|\\']", "", t)
    i = gsub("\\'", "", n)
    taxa.split = data.frame(cbind(i))
  }
  return(taxa.split)
}

# First input file
taxa.may17 = infile_may17$taxomomy
infile_may17 = infile_may17[,!(colnames(infile_may17)%in% 'taxomomy')]

taxa.split.may = format_taxonomy(taxa.may17)
taxa.split.may = data.frame(taxa.split.may)

# Second input file
taxa.oct17 = infile_oct17$taxomomy
infile_oct17 = infile_oct17[,!(colnames(infile_oct17)%in% 'taxomomy')]

taxa.split.oct = format_taxonomy(taxa.oct17)
taxa.split.oct = data.frame(taxa.split.oct)
```

```{r taxa_output, include=TRUE}
# Writing taxonomy outputs
write.csv(taxa.split.may, "R/CO1DvMay17_taxonomy.csv")
write.csv(taxa.split.oct, "R/CO1DvOct17_taxonomy.csv")
```

Each initial input file was generated as the result of two separate taxonomic assignment operations. First DNA centroids were assigned against reference database, the DNA reads that resulted unassigned were then assigned against the NCBI nucleotide (nt) database (available at: `ftp://ftp.ncbi.nlm.nih.gov/blast/db/` - updated to early September 2018).
Each operation generated two separate files which were then merged. The following step is to remove potential duplicate taxa records that would prevent R from definying the row names with the taxa names. Because we removed the last column with the taxonomy information, we can use the whole frame; otherwise we would have needed to avoid the last column in each frame.

```{r removing duplicates}
dv.may17 = as.data.frame(infile_may17 %>% group_by(X.OTU.ID) %>% summarise_all(funs(sum)))
dv.oct17 = as.data.frame(infile_oct17 %>% group_by(X.OTU.ID) %>% summarise_all(funs(sum)))
```

Considering the sequencing runs included different experiments, we exclude the columns containing headers belonging to other experiments. In this case `CH`,`Dv5`,`Ha`, and `pl4`.

```{r extracting_Dv-samples}
rownames(dv.may17) = dv.may17$X.OTU.ID
rownames(dv.oct17) = dv.oct17$X.OTU.ID

dv.may17 = dv.may17[,!(colnames(dv.may17)%in% 'X.OTU.ID')]
dv.oct17 = dv.oct17[,!(colnames(dv.oct17)%in% 'X.OTU.ID')]

to_remove1 = grep("Ha|CH|Dv5", colnames(dv.may17))
dv.may17 = dv.may17[,-to_remove1]

to_remove2 = grep("pl4|CH", colnames(dv.oct17))
dv.oct17 = dv.oct17[,-to_remove2]
```

# 1. Initial Quality Control

To calculate sample read depth for both input files, in order to quality check the overall sequencing, first all samples that didn't sequence at all (sample N read equal to 0) need to be removed. Then the mean and standard deviation for each data set can be calculated, and quality check of the samples read depth applied. 

```{r initial read depth, include=TRUE}
dv.may17 = subset(dv.may17, select = colSums(dv.may17) > 0)
dv.oct17 = subset(dv.oct17, select = colSums(dv.oct17) > 0)

may.read.depth = as.data.frame(dv.may17)
oct.read.depth = as.data.frame(dv.oct17)

may.read.depth$time = "May"
oct.read.depth$time = "Oct"

# Sum of DNA reads by samples to plot the sample read depth

may.read.depth = melt(may.read.depth)
may.read.depth = may.read.depth %>% group_by(time, variable) %>% summarise_all(funs(sum))

oct.read.depth = melt(oct.read.depth)
oct.read.depth = oct.read.depth %>% group_by(time, variable) %>% summarise_all(funs(sum))

# Joining both frames, then plotting

full.read.depth = rbind(may.read.depth, oct.read.depth)

ggplot(full.read.depth, aes(time, value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color=c("gray30"), alpha=0.5) +
  theme_bw() + labs(x="Data sets", y="Sample read depth (N reads)")
```

The sample N reads appear to be highly dispersed, this causes the mean-sd values to result negative. After checking for normality and log-transforming the data sets, the read depth threshold could be applied.


```{r mean and sd calculation}
mean(colSums(dv.may17))-sd(colSums(dv.may17))
mean(colSums(dv.oct17))-sd(colSums(dv.oct17))

shapiro.test(colSums(dv.may17))
shapiro.test(colSums(dv.oct17))

may.depth.thre = mean(log10(colSums(dv.may17)))-2*sd(log10(colSums(dv.may17)))
oct.depth.thre = mean(log10(colSums(dv.oct17)))-2*sd(log10(colSums(dv.oct17)))
```

The filter coverage for all samples that are equal and greater than mean-2sd of the log10-transformed coverage per samples were retained. The survival of the N samples was quite high in both data sets, with respectivelty 202 samples (93.5%) passing the read depth filter in May data set; and 186 samples (93.9%) in October data set.

```{r test_coverage, include=TRUE, echo=FALSE}
# Applying the threshold
reads.may17 = subset(dv.may17, select = log10(colSums(dv.may17)) >= may.depth.thre)
reads.oct17 = subset(dv.oct17, select = log10(colSums(dv.oct17)) >= oct.depth.thre)

# Checking N samples retained
count(log10(colSums(dv.may17)) <= may.depth.thre)[1,2]
count(log10(colSums(dv.oct17)) <= oct.depth.thre)[1,2]
count(log10(colSums(dv.may17)) <= may.depth.thre)[1,2]/(count(log10(colSums(dv.may17)) > 0)[,2])
count(log10(colSums(dv.oct17)) <= oct.depth.thre)[1,2]/(count(log10(colSums(dv.oct17)) > 0)[,2])
```

Unwanted taxa will now need to be removed before the contamination control.

```{r cleaning_frames}
# Cleaning empty taxa
reads.may17 = subset(reads.may17, subset = rowSums(dv.may17) > 0)
reads.oct17 = subset(reads.oct17, subset = rowSums(dv.oct17) > 0)

disc.may = c("Abramis_brama","Anser","Aphanomyces","Arthropoda_environmental_sample","Chromulinales","Cyclostephanos_dubius","Cyclostephanos_invisitatus","Cyclostephanos_sp._WTC16","Didymella_pinodes","Dinophyceae","Encyonema","Eukaryota","Fistulifera_solaris","Fusarium","Gomphonema_parvulum","Gymnocephalus","Hominidae","Homo_sapiens","Ilyonectria_destructans","invertebrate_environmental_sample","Lagenidium","Leptolegnia_sp._BOLD:AAX5717","Mallomonas_sp.","Melosira_ambiqua","Metazoa","Mus_musculus","Oncorhynchus_mykiss","Oomycetes","Paraphysomonas_sp.","Penicillium_sclerotiorum","Phytophthora","Pinnularia_neomajor","Pseudorasbora_parva","Pythium","Pythium_marsipium","Saprolegnia_delica","Saprolegniaceae","Sellaphora_obesa","Skeletonema_potamos","Sordariomycetes","Stephanodiscus","Stephanodiscus_cf._minutulus","Sus_scrofa","Tetracladium","Tetradesmus_obliquus","Thalassiosira_pseudonana","unassigned","Yamagishiella_unicocca","zooplankton_environmental_sample")

clean.may = reads.may17[!(rownames(reads.may17) %in% disc.may),]

disc.oct = c("Abramis_brama","Achlya","Actinopteri","Albugo_candida","Anatidae","Anser_anser","Aphanomyces","Apoikiospumella_mondseeiensis","Ascomycota","Aves","Batrachospermaceae","Batrachospermum_gelatinosum","Bodo_saltans","Branta_canadensis","Chlorella_sp._ArM0029B","Chroicocephalus_ridibundus","Chromulinales","Chrysochromulina","Chrysochromulina_parva","Chrysophyceae_sp.","Cochliopodium","Colletotrichum","Cordylophora_sp._IM1","Cordylophora_sp._NFR3","Cristatella_mucedo","Cyclostephanos_dubius","Cyclostephanos_invisitatus","Cyclostephanos_sp._WTC16","Cygnus_olor","Cylindrodendrum_hubeiense","Didymella_pinodes","Didymium_minus","Dinobryon","Dinobryon_pediforme","Dinophyceae","Emericellopsis","Eukaryota","Fistulifera_solaris","Fulica_atra","Fungi","Fusarium","Gomphonema_parvulum","Gymnocephalus_cernua","Heterolepidoderma_macrops","Homo_sapiens","Hypocreales","Ilyonectria_destructans","invertebrate_environmental_sample","Mallomonas_sp.","Melampsora","Melosira_ambiqua","Mesostigma_viride","Metazoa","Microcystis","Microcystis_aeruginosa","Mucorales","Myodes_glareolus","Nitzschia_palea","Nusuttodinium","Oncorhynchus_mykiss","Oomycetes","Ostreopsis","Paraphysomonas_sp.","Pedospumella","Penicillium","Percidae","Phytophthora","Phytophthora_gallica","Phytophthora_plurivora","Plumatella_fungosa","Podiceps_cristatus","Prunella_modularis","Pseudomuriella","Pseudoperonospora_humuli","Pycnococcaceae","Pythiaceae","Pythium","Pythium_marsipium","Pythium_phragmitis","Pythium_sp._JN-1","Ripella","Rutilus_rutilus","Salmo_trutta","Saprolegnia_delica","Scardinius","Segregatospumella_dracosaxi","Sellaphora_bacillum","Sellaphora_lanceolata","Sellaphora_obesa","Sellaphora_pupula","Sheathia_arcuata","Skeletonema","Spumella_sp.","Spumella_sp._LG-2014b","Stephanodiscus","Streptophyta","Tetracladium","Tetracladium_setigerum","Tetradesmus_obliquus","Thalassiosira_pseudonana","Thelonectria_discophora","Thelonectria_veuillotiana","Turdus_merula","unassigned","uncultured_Jaminaea","uncultured_Philodina","uncultured_Pythium","Vermamoeba_vermiformis","Verticillium","Verticillium_dahliae","zooplankton_environmental_sample","Zymoseptoria_tritici")

clean.oct = reads.oct17[!(rownames(reads.oct17) %in% disc.oct),]
```

The contamination control is based maninly on the `POSITIVE` samples. Those included single-specie tissue DNA of _Osmia bicornis_ (Megachilidae); and the proportion of other taxa detected in those single-specie samples will be the base on which the contamination threshold will be calculated.

```{r contamination control}
pos.may = clean.may[,grep("POS", colnames(clean.may))]
pos.oct = clean.oct[,grep("POS", colnames(clean.oct))]

colSums(pos.may)
colSums(pos.oct)

# We remove all taxa equal to 0

pos.may = subset(pos.may, subset = rowSums(pos.may) > 0)
pos.oct = subset(pos.oct, subset = rowSums(pos.oct) > 0)

# We decided to ignore one of the positive samples from May 2017 as it contained only 22 reads, of which none belonged to positive taxa _Osmia bicornis_

pos.may = subset(pos.may, select = colSums(pos.may) > 22)

colSums(pos.may["Osmia_bicornis",])
colSums(pos.oct["Osmia_bicornis",])

colSums(pos.may)-colSums(pos.may[c("Osmia_bicornis","Hymenoptera"),])
colSums(pos.oct)-colSums(pos.oct[c("Osmia_bicornis","Hymenoptera"),])

(colSums(pos.may)-colSums(pos.may[c("Osmia_bicornis","Hymenoptera"),]))/colSums(pos.may)
(colSums(pos.oct)-colSums(pos.oct[c("Osmia_bicornis","Hymenoptera"),]))/colSums(pos.oct)
```

Three of the positive samples in the October data set contained high amount of contamination from _Crangonyx floridanus_, and _Dikerogammarus villosus_ DNA. 
Specifically 648 reads (46.3%) belonging to _C. floridanus_ were present in a single positive sample; and 2106 reads (68.3%) and 2681 reads (71.6%) belonging to _D. villosus_ in two separate positive samples. In all these cases the proportion of reads was going to be too high to be used as threshold. The decision taken was then to exclude those three samples before calculating the contamination threshold for the October data set.

```{r high levels taxa}
pos.oct["Crangonyx_floridanus",]
pos.oct["Crangonyx_floridanus",]/colSums(pos.oct)
pos.oct["Dikerogammarus_villosus",]
pos.oct["Dikerogammarus_villosus",]/colSums(pos.oct)

# Removing the high contaminated positive samples
pos.oct = pos.oct[,-c(1:3)]
```

Based on the levels of contamination in the Positive samples, the threshold levels were decided to be respectively 0.03% (May data set) and 0.5% (October data set).

```{r threshold}
# Transposing the data frames and copying them to carry them over as N reads and ratios.
clean.may = as.data.frame(t(clean.may))

filter.may = as.data.frame(clean.may)
filter.may = filter.may[,]/rowSums(filter.may)
filter.may[is.na(filter.may)] <- 0

filter.may[][filter.may <= 0.0003] <- 0
clean.may[][clean.may[,]/rowSums(clean.may) <= 0.0003] <- 0

clean.oct = as.data.frame(t(clean.oct))

filter.oct = as.data.frame(clean.oct)
filter.oct = filter.oct[,]/rowSums(filter.oct)
filter.oct[is.na(filter.oct)] <- 0

filter.oct[][filter.oct <= 0.005] <- 0
clean.oct[][clean.oct[,]/rowSums(clean.oct) <= 0.005] <- 0

# Removing empty taxa

clean.may = subset(clean.may, select = colSums(clean.may) > 0)
filter.may = subset(filter.may, select = colSums(filter.may) > 0)

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)
```


# Remaining known contaminations

Checking for remaining contamination, _Harmonia axyridis_ is still present in one samples of May data set (RB501) with 7 reads (3.9%). Similarly, _Osmia bicornis_ is still present in few samples; however excluding the POSITIVE samples, the remaining contamination is in very low amount. Despite the contamination filter threshold didn't manage to remove these issues, the DNA read count is low enough to allow removal without major changes in the final output. Similar situation happened for the Primer contamination labeled `cont.13` which managed to pass the initial QC threshold despite having 17 reads (11 belonging to _O. bicornis_ and 6 belonging to _D. villosus_). The read count is low enough to feel confident in not having an influence in the overal output, therefore it can be excluded.

```{r remaining contaminations May}
# Checking in which sample(s) _Osmia bicornis_ contamination is still present
count(filter.may$Osmia_bicornis > 0)
rownames(filter.may)[filter.may$Osmia_bicornis > 0]

# Checking N reads of _O. bicornis_
clean.may$Osmia_bicornis[rownames(clean.may) %in% rownames(filter.may)[filter.may$Osmia_bicornis > 0]]

clean.may = clean.may[,!(colnames(clean.may) %in% "Osmia_bicornis")]
filter.may = filter.may[,!(colnames(filter.may) %in% "Osmia_bicornis")]

# Checking in which sample(s) _Harmonia axyridis_ contamination is still present
count(filter.may$Harmonia_axyridis > 0)
rownames(filter.may)[filter.may$Harmonia_axyridis > 0]

# Checking N reads of _H. axyridis_ 
clean.may$Harmonia_axyridis[rownames(clean.may) %in% rownames(filter.may)[filter.may$Harmonia_axyridis > 0]]

# Removing _H. axyridis_ without need of further filtration
clean.may = clean.may[,!(colnames(clean.may) %in% "Harmonia_axyridis")]
filter.may = filter.may[,!(colnames(filter.may) %in% "Harmonia_axyridis")]

# Checking extra contamination sample(s) still present
sum(clean.may["cont.13",])
colnames(filter.may)[filter.may["cont.13",] > 0]

# Checking N reads of _D. villosus_ and _O. bicornis_ present in pcr contamination sample
clean.may["cont.13","Dikerogammarus_villosus"]
clean.may["cont.13","Osmia_bicornis"]

# Removing `cont.13` sample without need of further filtration
clean.may = clean.may[!(rownames(clean.may) %in% "cont.13"),]
filter.may = filter.may[!(rownames(filter.may) %in% "cont.13"),]

clean.may = clean.may[-c(grep("BL", rownames(filter.may))),]
filter.may = filter.may[-c(grep("BL", rownames(filter.may))),]
```

In the samples from October, the contamination situation was less problematic. No _Harmonia axyridis_ remained after filter threshold; and _Osmia bicornis_ remained in only one samples (eGW4) with 28 reads (1.88%). In addition it was present contamination from _Porcellio scaber_ in two samples (RB409 and RB412), respectively at 2656 reads (5.6%), and 6234 reads (16.9%). Ultimately Psocoptera DNA was detected as contaminant in one sample (RB411), at 293 reads (0.9%).


```{r remaining contaminations Oct}
# Checking in which sample(s) _Osmia bicornis_ contamination is still present
count(filter.oct$Osmia_bicornis > 0)
rownames(filter.oct)[filter.oct$Osmia_bicornis > 0]

# Checking N reads of _O. bicornis_
clean.oct$Osmia_bicornis[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$Osmia_bicornis > 0]]

# Checking in which sample(s) _Psocoptera_ contamination is still present
count(filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0)
rownames(filter.oct)[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0]

# Checking N reads of _Psocoptera_
clean.oct$`Psocoptera_sp._BOLD:AAN8452`[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0]]
filter.oct[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0,"Psocoptera_sp._BOLD:AAN8452"]

# Checking in which sample(s) _Porcellio scaber_ contamination is still present
count(filter.oct$Porcellio_scaber > 0)
rownames(filter.oct)[filter.oct$Porcellio_scaber > 0]

# Checking N reads of _P. scaber_
clean.oct$Porcellio_scaber[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$Porcellio_scaber > 0]]
filter.oct[filter.oct$Porcellio_scaber > 0,"Porcellio_scaber"]

# Removing contamination
clean.oct = clean.oct[,!(colnames(clean.oct) %in% c("Osmia_bicornis","Psocoptera_sp._BOLD:AAN8452","Porcellio_scaber"))]
filter.oct = filter.oct[,!(colnames(filter.oct) %in% c("Osmia_bicornis","Psocoptera_sp._BOLD:AAN8452","Porcellio_scaber"))]

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)

clean.oct = subset(clean.oct, subset = rowSums(clean.oct) > 0)
filter.oct = subset(filter.oct, subset = rowSums(filter.oct) > 0)
```

The DNA extraction blanks (`GW.BL1 & 2`, `WB.BL1 & 2`, `RB.BL1 & 2`) resulted containing quite a high amount of contaminant DNA reads:

* `GW.BL1` = 3399, `GW.BL2` = 1782
* `WB.BL1` = 6338, `WB.BL2` = 863
* `RB.BL1` = 1818, `RB.BL2` = 2002

Of these, between 99.4% and 100% belonged to _Dikerogamamrus villosus_. This is far less than ideal considering those are meant to be dissection and extraction blanks; however the rest of the non-invaded Rockland Broad (RB) gut content samples presented very low _D. villosus_ contamination. In this perspective the decision was made not to act upon these levels of contamination also considering that _D. villosus_ is the main target taxa and overall in high concentration.

```{r extraction blanks}
# cleaning contamination in extraction blank samples
rowSums(clean.oct[grep("BL", rownames(filter.oct)),])
clean.oct[grep("BL", rownames(filter.oct)),"Dikerogammarus_villosus"]/rowSums(clean.oct[grep("BL", rownames(filter.oct)),])

clean.oct[grep("RB", rownames(filter.oct)),"Dikerogammarus_villosus"]

clean.oct = clean.oct[-c(grep("BL", rownames(filter.oct))),]
filter.oct = filter.oct[-c(grep("BL", rownames(filter.oct))),]

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)

clean.oct = subset(clean.oct, subset = rowSums(clean.oct) > 0)
filter.oct = subset(filter.oct, subset = rowSums(filter.oct) > 0)
```

Positive, Negative and Blank samples need to be removed before continuing.

```{r removing samples}
pnmay = grep("POS|NEG", rownames(filter.may), ignore.case=T)
filter.may = filter.may[-pnmay,]
clean.may = clean.may[-pnmay,]

pnoct = grep("POS|NEG", rownames(filter.oct), ignore.case=T)
filter.oct = filter.oct[-pnoct,]
clean.oct = clean.oct[-pnoct,]
```

Other variables need to be added to the data frames; specifically the sampling season, the gut contents samples ID, the site ID, and the predator ID (eDNA samples were labeled as "water"). Then the frames can be divided between gut contents and eDNA samples.

```{r adding variables}
# Adding time variable
filter.may$time = "May"
clean.may$time = "May"

filter.oct$time = "Oct"
clean.oct$time = "Oct"

# Adding sample ID from rownames
# May
filter.may$sample_id = gsub("[.]", "", rownames(filter.may))
clean.may$sample_id = gsub("[.]", "", rownames(clean.may))

# October
filter.oct$sample_id = gsub("[.]pl[0-9]{1}.{5,6}nc.blast", "", rownames(filter.oct))
clean.oct$sample_id = gsub("[.]pl[0-9]{1}.{5,6}nc.blast", "", rownames(clean.oct))

# Nested gsub to keep only the 2 letter from each sample ID referring to the lake ID
# May
filter.may$site = gsub("BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", filter.may$sample_id))))
clean.may$site = gsub("BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", clean.may$sample_id))))

# October
filter.oct$site = gsub("[.]BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", filter.oct$sample_id))))
clean.oct$site = gsub("[.]BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", clean.oct$sample_id))))

# Adding the sample type (eDNA or gut contents)
# May
ednamay = grep("e", filter.may$sample_id)
filter.may$sample_type = "NA"
filter.may[ednamay,"sample_type"] = "eDNA"
filter.may[-ednamay,"sample_type"] = "gut"

ednamay = grep("e", clean.may$sample_id)
clean.may$sample_type = "NA"
clean.may[ednamay,"sample_type"] = "eDNA"
clean.may[-ednamay,"sample_type"] = "gut"

# October
ednaoct = grep("e", filter.oct$sample_id)
filter.oct$sample_type = "NA"
filter.oct[ednaoct, "sample_type"] = "eDNA"
filter.oct[-ednaoct, "sample_type"] = "gut"

ednaoct = grep("e", clean.oct$sample_id)
clean.oct$sample_type = "NA"
clean.oct[ednaoct, "sample_type"] = "eDNA"
clean.oct[-ednaoct, "sample_type"] = "gut"
```

For the `predator_id`, the taxa ID was extracted from the taxa with ratio greater or equal to 90% within each samples, for both May and October. The samples that resulted still as "NA" at the end of this process were manually checked and the taxa ID were added manually.

```{r adding predator}
# Adding predator ID based on the DNA ratio greater than 0.9
# May
ednamay = grep("e", rownames(filter.may))
predmay = which(filter.may[-ednamay,!c(colnames(filter.may) %in% c("time","site","sample_id","sample_type"))] >= 0.9, arr.ind=T)
#rownames(predmay)
#colnames(filter.may[rownames(predmay), as.vector(predmay[,2])])

filter.may$predator_id = "NA"
filter.may[ednamay,"predator_id"] = "water"
filter.may[rownames(predmay),"predator_id"] = gsub("[.]?[0-9]{1,3}$", "", colnames(filter.may[rownames(predmay), as.vector(predmay[,2])]))

# October
ednaoct = grep("e", rownames(filter.oct))
predoct = which(filter.oct[-ednaoct,1:74] >= 0.9, arr.ind=T)
#rownames(predoct)
#colnames(filter.oct[rownames(predoct), as.vector(predoct[,2])])

filter.oct$predator_id = "NA"
filter.oct[ednaoct,"predator_id"] = "water"
filter.oct[rownames(predoct),"predator_id"] = gsub("[.]?[0-9]{1,3}$", "", colnames(filter.oct[rownames(predoct),as.vector(predoct[,2])]))

# Checking for missing predators
# May
count(filter.may$predator_id == "NA")
miss.may = gsub("[.]", "", rownames(filter.may[filter.may$predator_id == "NA",]))
miss.may

filter.may$predator_id[filter.may$sample_id %in% "WB507"] = "Dikerogammarus_villosus"
filter.may$predator_id[filter.may$sample_id %in% c("RB304","RB303")] = "Corophium_multisetosum"
filter.may$predator_id[filter.may$sample_id %in% c(miss.may[!miss.may %in% c("WB507","RB304","RB303")])] = "Gammarus_zaddachi"
clean.may$predator_id = filter.may$predator_id

# October
count(filter.oct$predator_id == "NA")
miss.oct = rownames(filter.oct[filter.oct$predator_id == "NA",])
miss.oct

filter.oct$predator_id[filter.oct$sample_id %in% c("GW103","GW108","GW401","GW601","WB627")] = "Dikerogammarus_villosus"
filter.oct$predator_id[filter.oct$sample_id %in% c("RB301","RB302","RB303","RB304","RB305","RB606","RB613","RB616")] = "Gammarus_zaddachi"
filter.oct$predator_id[filter.oct$sample_id %in% c("RB403","RB412")] = "Crangonyx_pseudogracilis"
clean.oct$predator_id = filter.oct$predator_id
```

Now that all variable have been added, the ratio and DNA reads of predators can be set to be equal to 0. This will allow further downstream to calulate accurate indexes of prey diversity and diet breadth without the risk of overinflating.

```{r removing predator ratio}
# Removal of predator ratios and DNA reads
# May
m.row = rownames(filter.may[filter.may$predator_id != "water", c("site","sample_id","time","sample_type","predator_id")])
m.col = filter.may[filter.may$predator_id != "water", "predator_id"]

filter.may[m.row,unique(m.col)]

for (i in m.row){
  pred_id = filter.may[i,"predator_id"]
  pred_read = clean.may[i, "predator_id"]
  filter.may[i,pred_id] <- 0
  clean.may[i,pred_read] <- 0
}

# Confirming the function worked
filter.may[m.row,unique(m.col)]

# Oct
o.row = rownames(filter.oct[filter.oct$predator_id != "water", c("site","sample_id","time","sample_type","predator_id")])
o.col = filter.oct[filter.oct$predator_id != "water", "predator_id"]

filter.oct[o.row,unique(o.col)]

for (i in o.row){
  pred_id = filter.oct[i,"predator_id"]
  pred_read = clean.oct[i, "predator_id"]
  filter.oct[i,pred_id] <- 0
  clean.oct[i,pred_read] <- 0
}

# Confirming the function worked
filter.oct[o.row,unique(o.col)]
```

Now that predators DNA ratio have been removed, the focus can move towards the prey id. Both _Crangonyx pseudogracilis_ and _C. floridanus_ appear detected in the data set; however despite recent records of this latter, its detection in the lakes included in this study couldn't be confirmed. The DNA reads belonging to those 2 species have been joined together. 

```{r}
clean.may$Crangonyx = clean.may$Crangonyx_floridanus + clean.may$Crangonyx_pseudogracilis
clean.oct$Crangonyx = clean.oct$Crangonyx_floridanus + clean.oct$Crangonyx_pseudogracilis
```

Saving the files into ".csv"

```{r saving files}
write.csv(clean.may, "R/May_reads_filtered.csv")
write.csv(filter.may, "R/May_ratio_filtered.csv")
write.csv(clean.oct, "R/Oct_reads_filtered.csv")
write.csv(filter.oct, "R/Oct_ratio_filtered.csv")
```

# 2. Detection of prey DNA in gut contents

Dividing gut contents and eDNA samples.

```{r dividing samples}
clean.guts = full_join(clean.may[clean.may$sample_type == "gut",], clean.oct[clean.oct$sample_type == "gut",])
clean.guts[is.na(clean.guts)] <- 0

clean.edna = full_join(clean.may[clean.may$sample_type == "eDNA",], clean.oct[clean.oct$sample_type == "eDNA",])
clean.edna[is.na(clean.edna)] <- 0
```

```{r removing empty taxa}
info.guts = clean.guts[,c(colnames(clean.guts) %in% c("time","site","sample_id","sample_type","predator_id"))]
info.edna = clean.edna[,c(colnames(clean.edna) %in% c("time","site","sample_id","sample_type","predator_id"))]

# Cleaning empty taxa
# gut content samples
clean.guts = subset(clean.guts[,!c(colnames(clean.guts) %in% colnames(info.guts))], select=colSums(clean.guts[,!c(colnames(clean.guts) %in% colnames(info.guts))]) > 0)
# eDNA samples
clean.edna = subset(clean.edna[,!c(colnames(clean.edna) %in% colnames(info.edna))], select=colSums(clean.edna[,!c(colnames(clean.edna) %in% colnames(info.edna))]) > 0)
```

The predator DNA reads now are equal to 0 across all gut content samples. This allows for calculating the prey DNA ratio within each gut content sample, that will allow also for a niche overlap analysis.
the prey species detected included also both _Crangonyx pseudogracilis_ and _C. floridanus_. Becuase _C.floridanus_ has been detected in the UK, but hasn't been confirmed in Rockland Broad, the decision was taken to be conservative and pull both Crangonyx species up to the genus level summing their DNA reads.
Furthermore the taxa `Isopoda`, that should have been assigned to _Asellus aquaticus_, can't really be used for our purposes because of our detected contamination from _Porcellio scaber_. The 1217 read belonging to Isopoda will need to be removed from the dataset.

```{r}
clean.guts$Crangonyx = clean.guts$Crangonyx_floridanus + clean.guts$Crangonyx_pseudogracilis
clean.guts = clean.guts[,!c(colnames(clean.guts) %in% c("Crangonyx_pseudogracilis","Crangonyx_floridanus"))]
clean.guts = clean.guts[,!c(colnames(clean.guts) %in% "Isopoda")]

ratio.guts = clean.guts[,]/rowSums(clean.guts)
ratio.guts[is.na(ratio.guts)] <- 0

ratio.guts = subset(ratio.guts, rowSums(ratio.guts) > 0)

# Using row names to subset the info associated with the gut contents
rownames(ratio.guts)
info.guts = info.guts[rownames(ratio.guts),]
```

The miss-identification of few predators is reflected in the results of the sequencing on the gut contents. So far all predators were included to allow for accurate quality controls; however before continuing into more semi-quantitative analysis of overlap and diversity indexes those mis-identified predators will need to be removed. 11 samples corresponding to _Corophium multisetosum_ (N=2), _Gammarus tigrinus_ (N=2), and _Crangonyx pseudogracilis_ (N=7) will need to be removed from the dataset as they won't provide accurate comparisons with the rest of the dataset.

```{r prey ratio in guts}
ratio.guts = cbind(ratio.guts, info.guts)
ratio.guts = ratio.guts[ratio.guts$predator_id %in% c("Dikerogammarus_villosus","Gammarus_zaddachi"),]
melt.guts = melt(ratio.guts)
```

```{r initial plot}
ggplot(melt.guts, aes(time, value)) +
  geom_jitter(aes(color=predator_id)) +
  facet_grid (.~site) + theme_bw() +
  ylab("prey DNA ratio") + xlab("sampling month")
```



```{r prey detection rarefaction curves}
# removing `sample_type` and `sample_id` variables and transforming detections into abundances
ratio.guts = ratio.guts[,!c(colnames(ratio.guts) %in% c("sample_type","sample_id"))]
ratio.guts[,!c(colnames(ratio.guts) %in% c("time","site","predator_id"))][ratio.guts[,!c(colnames(ratio.guts) %in% c("time","site","predator_id"))] > 0] <- 1

# rarefaction curves
abund.guts = as.data.frame(ratio.guts %>% group_by(time, site, predator_id) %>% summarise_all(funs(sum)))

# replacing predator id with initials
abund.guts[grep("Gammarus_zaddachi", abund.guts$predator_id),"predator_id"] <- "G.zad"
abund.guts[grep("Dikerogammarus_villosus", abund.guts$predator_id),"predator_id"] <- "D.vil"

# setting rownames and removing extra variables
rownames(abund.guts) = paste(abund.guts$predator_id, abund.guts$site, abund.guts$time, sep="_")
abund.guts = abund.guts[,!c(colnames(abund.guts) %in% c("site","time","predator_id"))]
abund.guts = t(abund.guts)

guts_inext = iNEXT(abund.guts, q=c(0,1), datatype = "abundance", se=T, endpoint=200)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(guts_inext) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Species level rarefaction curves of preys in gut contents")
```

# 3. Diversity analysis of eDNA samples plus comparison with gut contents
To avoid over-estimating the diversity present in the eDNA samples, the data set need to be cleaned a bit further. For this purpose assignment at higher taxonomic rank (mainly Order), were removed from the dataset, while assignments to Family, were spearated from the assignments to Genus and Species level. 

```{r eDNA samples rarefaction curves}
clean.edna = clean.edna[,!c(colnames(clean.edna) %in% c("Cyclopoida","Diptera","Podocopida","Spongillida"))]

# splitting Family ranks from Genus and Species ranks
fam.edna = clean.edna[,c(colnames(clean.edna) %in% c("Chironomidae","Chydoridae","Cyclopidae","Daphniidae","Naididae"))]

gen.sp.edna = clean.edna[,!c(colnames(clean.edna) %in% c("Chironomidae","Chydoridae","Cyclopidae","Daphniidae","Naididae"))]

# calculating ratio
gen.sp.edna = gen.sp.edna[]/rowSums(gen.sp.edna)
gen.sp.edna = cbind(gen.sp.edna, info.edna)

# removing `predator_id`, `sample_type` and `sample_id` variables and transforming eDNA detections into abundances
abund.edna = gen.sp.edna[,!c(colnames(gen.sp.edna) %in% c("sample_type","predator_id"))]
abund.edna[,!c(colnames(abund.edna) %in% c("time","sample_id","site"))][abund.edna[,!c(colnames(abund.edna) %in% c("time","sample_id","site"))] > 0] <- 1

# adding genus information to help in plotting
melt.abund.edna = melt(abund.edna)
melt.abund.edna$genus = gsub("([_]{1}[A-Za-z0-9-]{+}.{+}[.]{0,1}){0,+}$","", melt.abund.edna$variable)

# NOTE: to find the right way to plot this data! (barplot?) 
# Also alpha-div and raref. curves (need to transpose and sort sample names!) need to be done on abund.edna frame (genus column causes issues), while the plot of the data visualisation needs to be done on the melt.abund.edna so to summarise the species by the genus column.
# Next steps overlap between guts and eDNA. Next section kick-samples.

# Rarefaction curves calculated on abundance data without genus information
abund.edna = as.data.frame(abund.edna[,!c(colnames(abund.edna) %in% "sample_id")] %>% group_by(site, time) %>% summarise_all(funs(sum)))
rownames(abund.edna) = paste(abund.edna$site, abund.edna$time, sep="_")

# removing `time`, `sample_id` and `site` variables as no more necessary.
abund.edna = t(abund.edna[,!c(colnames(abund.edna) %in% c("time","site"))])

# extrapolating rarefaction curves for Species richness and Shannon Diversity
edna_inext = iNEXT(abund.edna, q=c(0,1), datatype = "abundance", se=T, endpoint=200)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(edna_inext) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Species level rarefaction curves of eDNA samples")
```

```{r eDNA and gut contents overlap}
View(abund.edna)
View(abund.guts)




## FROM HERE!!!!
```



# 3. Comparison between eDNA, kick-samples and gut contents detections








Once We can now start plotting and running models!

```{r}
gut.ratio = dvmay.filter[dvmay.filter$sample_type == "gut",]
gut.melt = melt(gut.ratio)

head(gut.melt)

gut.ratio2 = dvoct.filter[dvoct.filter$sample_type == "gut",]
gut.ratio2[3, "predator_id"] <- "Dikerogammarus_villosus"
gut.melt2 = melt(gut.ratio2)

gut.full = full_join(gut.melt, gut.melt2)

ggplot(gut.full, aes(time, value)) +
  geom_jitter(aes(color=predator_ID)) +
  facet_grid (.~site) +
  ylab("prey DNA ratio") + xlab("sampling month")
```

```{r}
niche.frame = full_join(gut.ratio, gut.ratio2)
niche.frame[is.na(niche.frame)] <- 0

niche.info = niche.frame[,c(colnames(niche.frame) %in% c("time", "sample_id","site","sample_type","predator_id"))]

niche.frame = subset(niche.frame[,!c(colnames(niche.frame) %in% c("time", "sample_id","site","sample_type","predator_id"))], select = colSums(niche.frame[,!c(colnames(niche.frame) %in% c("time", "sample_id","site","sample_type","predator_id"))]) > 0)

niche.frame[][niche.frame > 0] <- 1

niche.frame = cbind(niche.frame, niche.info)
```

# summing prey DNA reads by predator ID, site and time.

```{r}
colnames(niche.frame)
head(niche.frame)

nrow(niche.frame)

niche.model = as.data.frame(niche.frame[,!c(colnames(niche.frame) %in% c("sample_type","sample_id"))] %>% group_by(site, time, predator_ID) %>% summarise_all(funs(sum)))

# Becuase _C.f loridanus_ hasn't been confirmed in Rockland Broad, we decided to be conservative and pull both Crangonyx species up to the genus level.

niche.model$Crangonyx = niche.model$Crangonyx_pseudogracilis + niche.model$Crangonyx_floridanus
niche.model = niche.model[,!c(colnames(niche.model) %in% c("Crangonyx_pseudogracilis", "Crangonyx_floridanus"))]

ncol(niche.model)

niche.model = niche.model[-c(2,4,6),]
niche.model = niche.model[,!c(colnames(niche.model) %in% c("Dikerogammarus_villosus","Porcellio_scaber"))]

## test of the null model with ra2 algorithm
niche.null.model = niche_null_model(speciesData=niche.model[,4:ncol(niche.model)], algo="ra2", metric="pianka", suppressProg=T, nReps=5000)

summary(niche.null.model)
plot(niche.null.model,type="hist")
plot(niche.null.model,type="niche")

## test of the null model with ra3 algorithm
niche.null.model3 = niche_null_model(speciesData=niche.model[,4:ncol(niche.model)], algo="ra3", metric="pianka", suppressProg=T, nReps=5000)

summary(niche.null.model3)
plot(niche.null.model3,type="hist")

niche.model
rowSums(niche.model[,4:ncol(niche.model)])

## test of the null model with ra4 algorithm
#niche.null.model4 = niche_null_model(speciesData=niche.model[,4:ncol(niche.model)], algo="ra4", metric="pianka", suppressProg=T, nReps=5000)

#summary(niche.null.model4)
#plot(niche.null.model4,type="hist")

```

```{r}
head(niche.frame)

niche.nmds = as.data.frame(niche.frame)

niche.nmds = subset(niche.nmds, subset=rowSums(niche.nmds[,!c(colnames(niche.frame) %in% c("time","site","sample_id","predator_id","sample_type"))]) > 0)

matrix.niche = as.matrix(niche.nmds[,!c(colnames(niche.nmds) %in% c("time","site","sample_id","predator_id","sample_type"))])

fact_site.niche = matrix(c(niche.nmds$site),ncol=1) 
fact_time.niche = matrix(c(niche.nmds$time), ncol=1)
fact_predator.niche = matrix(c(niche.nmds$predator_ID), ncol=1)
fact_all.niche = matrix(c(niche.nmds$site, niche.nmds$time, niche.nmds$predator_ID), ncol=3)

dist.niche = vegdist(matrix.niche, method="euclidean")

adonis(dist.niche~fact_site.niche*fact_predator.niche, permutations = 10000, method="euclid")
adonis(dist.niche~fact_time.niche*fact_predator.niche, permutations = 10000, method="euclid")
```

```{r}
meta.niche = metaMDS(dist.niche, zerodist=ignore, trymax=200)
```

```{r}
plot(meta.niche,type="n",
main="gut contents PERMANOVA")#, ylim=c(-0.7,0.9), xlim=c(-0.5,1))
points(meta.niche, select=which(fact_site.niche[,1]=="WB"),col="black",pch=1)
points(meta.niche, select=which(fact_site.niche[,1]=="GW"),col="blue",pch=22)
points(meta.niche, select=which(fact_site.niche[,1]=="RB"),col="red",pch=25)
ordispider(meta.niche,group=fact_site.niche[,1], col=c("green","orange","green4"),label=T)
ordiellipse(meta.niche,group=fact_site.niche[,1],kind="ehull", lty=c(1,3),label=F,lwd=2)
```

```{r}
plot(meta.niche,type="n",
main="gut contents PERMANOVA")#, ylim=c(-0.7,0.9), xlim=c(-0.5,1))
points(meta.niche, select=which(fact_site.niche[,1]=="WB"),col="black",pch=1)
points(meta.niche, select=which(fact_site.niche[,1]=="GW"),col="blue",pch=22)
points(meta.niche, select=which(fact_site.niche[,1]=="RB"),col="red",pch=25)
ordispider(meta.niche,group=fact_predator.niche[,1],col=as.numeric(factor(fact_predator.niche[,1])),label=T)
ordiellipse(meta.niche,group=fact_time.niche[,1],kind="ehull",lty=c(1,3), label=F,lwd=2)

#,lty=as.numeric(factor(fact_predator.niche[,1]))
#c("green","orange","green4")
```

# Food webs

```{r}
niche.model

niche.web = niche.model[-c(2,4,6),]

rownames(niche.web) = c("Dv_GW-May","Dv_GW-Oct","Gz_RB-May","Gz_RB-Oct","Dv_WB-May","Dv_WB-Oct")

niche.web = niche.web[,!c(colnames(niche.web) %in% c("site","time","predator_id","Dikerogammarus_villosus","Porcellio_scaber"))]

niche.web = as.data.frame(t(niche.web))

str(niche.web)

niche.web = as.matrix(niche.web)
```

# NEED TO REMOVE Dv FROM WEB AND DIPTERA!!!!

```{r tripartite web, fig.keep="last"}
plotweb(niche.web, y.width.low=0.1, y.width.high=0.08,	y.lim=c(0,3), low.lablength=5, col.high="white", low.lab.dis=0.1, high.lab.dis=0.1, low.spacing=0.012, text.rot=90)
```


```{r end}
sink("log_session_packages.txt")
sessionInfo()
citation()
sink()
```

## END



# Testing for the main 2 predators _Dikerogammarus villosus_ (for WB and GW) and _Gammarus zaddachi_ (for RB).

```{r}
gut.full2 = subset(gut.full, subset=gut.full$value > 0)

test.glm = lm(gut.full2$value ~ gut.full2$site, family = quasibinomial)
test.glm2 = lm(gut.full2$value ~ gut.full2$predator_ID, family = quasibinomial)
summary(test.glm)
summary(test.glm2)

plot(test.glm)
plot(test.glm2)

test.gam = gam(gut.full$value[gut.full$predator_ID == c("Dikerogammarus_villosus","Gammarus_zaddachi")] ~ gut.full$site[gut.full$predator_ID == c("Dikerogammarus_villosus","Gammarus_zaddachi")] + gut.full$predator_ID[gut.full$predator_ID == c("Dikerogammarus_villosus","Gammarus_zaddachi")])

test.gam2 = gam(gut.full$value ~ gut.full$site + gut.full$predator_ID)

summary(test.gam)
summary(test.gam2)

gam.check(test.gam)
gam.check(test.gam2)

#gut.full2 %>%
#  mutate(variable = reorder(variable, value)) %>%
#  ggplot( aes(x=variable, y=value, fill=variable, color=variable)) +
#    geom_violin(width=2.1, size=0.2) +
#    scale_fill_viridis(discrete=TRUE) +
#    scale_color_viridis(discrete=TRUE) +
#    facet_free(site~time) +
#    theme_classic() +
#    theme(legend.position="none") +
#    coord_flip() +
#    xlab("") +
#    ylab("ratio(%)")
```

```{r}

edna.ratio = dvmay.filter[dvmay.filter$sample_type == "eDNA",]
edna.melt = melt(edna.ratio)
edna.melt = subset(edna.melt, subset=edna.melt$value > 0)

edna.ratio2 = dvoct.filter[dvoct.filter$sample_type == "eDNA",]
edna.melt2 = melt(edna.ratio2)
edna.melt2 = subset(edna.melt2, subset=edna.melt2$value > 0)

edna.full = full_join(edna.melt, edna.melt2)

head(edna.full)

ggplot(abund.edna, aes(sample_id, value)) +
  geom_jitter() +
  geom_bar(stat="identity", position="fill", aes(fill=variable)) +
  facet_grid (.~time) +
  xlab("sampling month")

```



```{r splitting samples}
guts.all = full_join(guts.may, guts.oct)
guts.all[is.na(guts.all)] <- 0

edna.all = full_join(edna.may, edna.oct)
edna.all[is.na(edna.all)] <- 0

chr = grep("samples_id|time", colnames(guts.all))
guts.melt = subset(guts.all, select = colSums(guts.all[,-chr]) > 0)
guts.melt$samples_id = guts.all$samples_id


chr2 = grep("samples_id|time", colnames(edna.all))
edna.melt = subset(edna.all, select = colSums(edna.all[,-chr2]) > 0)
edna.melt$time = edna.all$time

guts.melt2 = melt(guts.melt)
edna.melt2 = melt(edna.melt)

ggplot(guts.melt2, aes(samples_id, value)) +
  geom_bar(stat="identity", aes(fill=variable)) +
  facet_grid(time~.)
```


```{r}
## DIVIDING THE DATA FRAME BY LAKE ##
#guts.melt$site = gsub("[0-9]{1,3}","", guts.melt$samples_id)

head(guts.melt)
gw = grep("GW", guts.melt$samples_id, fixed=T)
guts.gw = guts.melt[gw,]

wb = grep("WB", guts.melt$samples_id, fixed=T)
guts.wb = guts.melt[wb,]

rb = grep("RB", guts.melt$samples_id, fixed=T)
guts.rb = guts.melt[rb,]

tssgw = grep("time|samples_id|site", colnames(guts.gw))
gw.melt = subset(guts.gw, select=colSums(guts.gw[,-tssgw]) > 0)
gw.melt = cbind(gw.melt, guts.gw[,tssgw])

tsswb = grep("time|samples_id|site", colnames(guts.wb))
wb.melt = subset(guts.wb, select=colSums(guts.wb[,-tsswb]) > 0)
wb.melt = cbind(wb.melt, guts.wb[,tsswb])

tssrb = grep("time|samples_id|site", colnames(guts.rb))
rb.melt = subset(guts.rb, select=colSums(guts.rb[,-tssrb]) > 0)
rb.melt = cbind(rb.melt, guts.rb[,tssrb])

guts.gw[,!(colnames(guts.gw) %in% c("time","samples_id","site"))]

### writing output to file
write.csv(guts.wb, "R/wroxham_gutcontents.csv")
write.csv(guts.gw, "R/grafham_gutcontents.csv")
write.csv(guts.rb, "R/rockland_gutcontents.csv")
```

## Model running
Before running any model we need to make sure that we decide correctly whether to change the the proportions to pres/abs (1,0). In order to keep the semi-quantification at population level (the semi-quantification at individual is outside of our scopes) we add an extra information regarding the predator identity (we will need to remove the predator variable from the ratio).

```{r}
colnames(guts.gw)


  



gw.melt = melt(guts.gw)
wb.melt = melt(guts.wb)
rb.melt = melt(guts.rb)

gw.melt$value[][gw.melt$value > 0] <- 1


## Removing the predator we detect a significant correlation (p=0.00123) between prey species and season

coplot(gw.melt$value[!(gw.melt$variable %in% "Dikerogammarus_villosus")] ~ gw.melt$variable[!(gw.melt$variable %in% "Dikerogammarus_villosus")] | gw.melt$time[!(gw.melt$variable %in% "Dikerogammarus_villosus")])
gw.model <- lm(gw.melt$value[!(gw.melt$variable %in% "Dikerogammarus_villosus")] ~ gw.melt$variable[!(gw.melt$variable %in% "Dikerogammarus_villosus")] * gw.melt$time[!(gw.melt$variable %in% "Dikerogammarus_villosus")])
summary(gw.model)
drop1(gw.model,test = "F")

plot(gw.model, add.smooth = FALSE, which = 1)
gwres <- resid(gw.model)
hist(gwres, xlab = "Residuals", main = "")
plot(gw.melt$value[!(gw.melt$variable %in% "Dikerogammarus_villosus")], gwres, xlab = "prey detection",
ylab = "Residuals")
boxplot(gwres ~ gw.melt$time[!(gw.melt$variable %in% "Dikerogammarus_villosus")], xlab = "Month",
ylab = "Residuals")
```

```{r}
wb.melt$value[][wb.melt$value > 0] <- 1


## Removing the predator we detect a significant correlation (p=0.00123) between prey species and season

coplot(wb.melt$value[!(wb.melt$variable %in% "Dikerogammarus_villosus")] ~ wb.melt$variable[!(wb.melt$variable %in% "Dikerogammarus_villosus")] | wb.melt$time[!(wb.melt$variable %in% "Dikerogammarus_villosus")])
wb.model <- lm(wb.melt$value[!(wb.melt$variable %in% "Dikerogammarus_villosus")] ~ wb.melt$variable[!(wb.melt$variable %in% "Dikerogammarus_villosus")] * wb.melt$time[!(wb.melt$variable %in% "Dikerogammarus_villosus")])
drop1(wb.model,test = "F")

plot(wb.model, add.smooth = FALSE, which = 1)
wbres <- resid(wb.model)
hist(wbres, xlab = "Residuals", main = "")
plot(wb.melt$value[!(wb.melt$variable %in% "Dikerogammarus_villosus")], wbres, xlab = "prey detection",
ylab = "Residuals")
boxplot(wbres ~ wb.melt$time[!(wb.melt$variable %in% "Dikerogammarus_villosus")], xlab = "Month",
ylab = "Residuals")
```

```{r}
rb.melt$variable[rb.melt$value > 0.75]
rb.melt$value[!(rb.melt$value > 0.9)]

## Removing the predator we detect a significant correlation (p=0.00123) between prey species and season

coplot(rb.melt$value[!(rb.melt$value > 0.75)] ~ rb.melt$variable[!(rb.melt$value > 0.75)] | rb.melt$time[!(rb.melt$value > 0.75)])
rb.model <- lm(rb.melt$value[!(rb.melt$value > 0.75)] ~ rb.melt$variable[!(rb.melt$value > 0.75)] * rb.melt$time[!(rb.melt$value > 0.75)])
drop1(rb.model,test = "F")

plot(rb.model, add.smooth = FALSE, which = 1)
rbres <- resid(rb.model)
hist(rbres, xlab = "Residuals", main = "")
plot(rb.melt$value[!(rb.melt$value > 0.75)], rbres, xlab = "prey detection",
ylab = "Residuals")
boxplot(rbres ~ rb.melt$time[!(rb.melt$value > 0.75)], xlab = "Month",
ylab = "Residuals")
```

```{r}


plot(glm(gw.melt$variable ~ gw.melt$samples_id*gw.melt$time, family = binomial))

anosim(guts.gw[,!(colnames(guts.gw) %in% c("time","samples_id","site"))], guts.gw$time, distance="euclid", permutations=1000)
metaMDS(as.matrix(vegdist(guts.gw[,!(colnames(guts.gw) %in% c("time","samples_id","site"))], distance="euclid")), distance="euclid", trymax = 1000)

anosim(guts.melt[,!(colnames(guts.melt) %in% c("time","samples_id","site"))], guts.melt$site, distance="euclid")
metaMDS(as.matrix(vegdist(guts.melt[,!(colnames(guts.melt) %in% c("time","samples_id","site"))], distance="euclid")), distance="euclid", trymax = 10000)
```

## OLD SCRIPT

```{r}
names(test)=ordered(names(test), levels=c("eWB1","eWB2","eWB3","eWB4","eWB5","eWB6",
                                                "eGW1","eGW2","eGW3","eGW4","eGW5","eGW6"))

test = melt(test)
names(test)
names(test)[1]="family"

g<-ggplot(test, aes(variable,value,fill=family))
g<-g+geom_bar(stat="identity", position="fill", width=0.75)
#g<-g+geom_boxplot()
g<-g+theme_bw()#scale_fill_brewer(palette="Set1")
g<-g+xlab("eDNA sampling sites") + 
ylab("Filtered DNA reads ratios")
g<-g+theme(axis.text = element_text(size=15)) + 
theme(axis.title = element_text(size=15))
g

## SAVE IMAGE ##

#ggsave("eDNA-coiAug-barplot.emf", width=8, height=10)

## TISSUE SAMPLES FOR THE TROPHIC WEB ##

tissue = read.csv("CO1Aug_gutcontent_filtered.csv", header=T, row.names=1)
head(tissue)

colSums(tissue)
rowSums(tissue)

## CLEANING FROM EMPTY SAMPLES AND SPECIES ##

dim(tissue)
tissue = subset(tissue, select = c(colSums(tissue) > 0))
dim(tissue)
tissue = subset(tissue, subsett = c(rowSums(tissue) > 0))
dim(tissue)

## DIVIDING THE DATA FRAME BY LAKE ##

tissue.plot = t(tissue)

gw = grep("GW", colnames(tissue.plot), fixed=T)
for (t in tissue.plot[,gw]){
	tissue.gw = data.frame(subset(tissue.plot[,gw]))
}
dim(tissue.gw)

wb = grep("WB", colnames(tissue.plot), fixed=T)
for (t in tissue.plot[,wb]){
	tissue.wb = data.frame(subset(tissue.plot[,wb]))
}
dim(tissue.wb)

## NEED TO CLEAN THE FRAMES FROM EMPTY CELLS ##
######
## Surely there's a loop that can be made, but for
## now the manual version will do the job
######

rowSums(tissue.gw)
rowSums(tissue.wb)

dim(tissue.gw)
tissue.gw = subset(tissue.gw, select=c(colSums(tissue.gw) > 0))
dim(tissue.gw)
tissue.gw = subset(tissue.gw, subset=c(rowSums(tissue.gw) > 0))
dim(tissue.gw)

dim(tissue.wb)
tissue.wb = subset(tissue.wb, select=c(colSums(tissue.wb) > 0))
dim(tissue.wb)
tissue.wb = subset(tissue.wb, subset=c(rowSums(tissue.wb) > 0))
dim(tissue.wb)

rownames(tissue.gw)
rownames(tissue.wb)
tissue.wb=tissue.wb[!(rownames(tissue.wb)%in% "Anser_anser"),]
############################################################
## CHANGING THE NAME OF SAMPLE BY ADDING NAME OF PREDATOR ##
############################################################

tissue.gw
g = which(tissue.gw >= 0.9, arr.ind=T)
rownames(g)
g[,2]
colnames(tissue.gw) = paste(rownames(g), colnames(tissue.gw[which(colnames(tissue.gw) == g[,2]),]), sep="_")
tissue.gw

tissue.wb
w = which(tissue.wb >= 0.7, arr.ind=T)
rownames(w)
w[,2]
colnames(tissue.wb) = paste(rownames(w), colnames(tissue.wb[which(colnames(tissue.wb) == w[,2]),]), sep="_")
tissue.wb

write.csv(tissue.gw, "coi-eDNA/tissue_gw.csv")
write.csv(tissue.wb, "coi-eDNA/tissue_wb.csv")

## SWAPPING ALL VALUES ABOVE 0 TO 1 ##

tissue.wb[,][tissue.wb[,] > 0] <- 1
tissue.gw[,][tissue.gw[,] > 0] <- 1

################################################
## SETTING UP THE GW AND WB WEBS FRAMES USING ##
## D. villosus GUT CONTENTS                   ##
################################################

colnames(tissue.wb)
colnames(tissue.gw)

wb.web = tissue.wb[!(rownames(tissue.wb)%in% "Dikerogammarus_villosus"),]
gw.web = tissue.gw[!(rownames(tissue.gw)%in% "Dikerogammarus_villosus"),]

wb.web$D_villosus_WB = rowSums(wb.web[,grep("Dikerogammarus_villosus", colnames(wb.web))])
wb.web = subset(wb.web, select = c(colnames(wb.web)%in% 'D_villosus_WB'))

gw.web$D_villosus_GW = rowSums(gw.web[,grep("Dikerogammarus_villosus", colnames(gw.web))])
gw.web = subset(gw.web, select = c(colnames(gw.web)%in% 'D_villosus_GW'))

## LOADING eDNA AND KICK-SAMPLE DATA FOR ESTIMATING ABUNDANCES ##

dir()
dna.water = read.csv("CO1Aug_water-samples_filtered.csv", header=T, row.names=1)

head(dna.water)
## dna.water$samples = substr(rownames(dna.water), 2, 
## nchar(rownames(dna.water))-1)

dna.water = t(dna.water)

## MAKING ALL VALUES > 0 EQUAL TO 1 PLUS ##
## EXTRACTING 'GW' AND 'WB' SAMPLES FROM THE FRAME ##

dna.water[1:nrow(dna.water),][dna.water[1:nrow(dna.water),] > 0] <- 1

colnames(dna.water)

gw.list = grep("GW", colnames(dna.water), fixed=T)
for (t in dna.water[,gw.list]){
	gw.water = data.frame(subset(dna.water[,gw.list]))
}
wb.list = grep("WB", colnames(dna.water), fixed=T)
for (t in dna.water[,wb.list]){
	wb.water = data.frame(subset(dna.water[,wb.list]))
}

## IMPORTING MICROSCOPY DATA ##

dir("coi-eDNA")
#micros = read.csv("coi-eDNA/ID_gw-wb.csv", header=T)
micros=read.csv("coi-eDNA/microscopy_gw-wb.csv", header=T, row.names=1)
micros
names(micros)
str(micros)
micros=data.frame(t(micros))

#rownames(micros)= micros$BINOMIAL

gwmic=micros[,grep("GW", colnames(micros))]
wbmic=micros[,grep("WB", colnames(micros))]

gwmic[1:nrow(gwmic),][gwmic[1:nrow(gwmic),] > 0] <- 1
wbmic[1:nrow(wbmic),][wbmic[1:nrow(wbmic),] > 0] <- 1

names(gwmic)
names(wbmic)
ncol(gwmic)
ncol(wbmic)
dim(gwmic)
dim(wbmic)

## JOINING AND CALCULATING TOTAL OCCUPANCY (COMBINING eDNA and MICROSCOPY) ##

rownames(gwmic)
rownames(gw.water)
dim(gw.water)
dim(gw.web)
gwmic
gwmic$species = rownames(gwmic)
gw.water$species = rownames(gw.water)

gw.joined = full_join(gwmic, gw.water, by="species")
gw.joined[is.na(gw.joined)] <- 0
gw.joined

sites = "GW1|GW2|GW3|GW4|GW5|GW6"
gw.joined$GW_occup = (rowSums(gw.joined[,grep(sites, colnames(gw.joined))])/6)
gw.joined

names(wbmic)
names(wb.water)
wbmic
wbmic$species = rownames(wbmic)
wb.water$species = rownames(wb.water)

wb.joined = full_join(wbmic, wb.water, by="species")
wb.joined[is.na(wb.joined)] <- 0
wb.joined

sites = "WB1|WB2|WB3|WB4|WB5|WB6"
wb.joined$WB_occup = (rowSums(wb.joined[,grep(sites, colnames(wb.joined))])/6)
wb.joined

##################################
## SETTING THE ESTIMATE VECTORS ##
##################################

gw.est = subset(gw.joined, select = c(colnames(gw.joined)%in% 'GW_occup'))
wb.est = subset(wb.joined, select = c(colnames(wb.joined)%in% 'WB_occup'))

gw.est
wb.est

gw.est = as.vector(gw.est[,1])
names(gw.est) = c(gw.joined$species)
names(gw.est)

wb.est = as.vector(wb.est[,1])
names(wb.est) = c(wb.joined$species)
names(wb.est)

###############################################
## PLOTTING THE WEBS WITH THE eDNA ESTIMATES ##
###############################################

#dev.new()
#par(mfrow=c(1,2))

str(wb.web)
str(wb.est)

write.csv(gw.web, "DV_foodwebs_2018/gw_web.csv")
write.csv(gw.est, "DV_foodwebs_2018/gw_est.csv")
          
write.csv(wb.web, "DV_foodwebs_2018/wb_web.csv")
write.csv(wb.est, "DV_foodwebs_2018/wb_est.csv")

## IMPROVING LABELS FOR PLOTS ##
gw.web = read.csv("DV_foodwebs_2018/gw_web.csv", row.names = 1, header=T)
gw.est = read.csv("DV_foodwebs_2018/gw_est.csv", row.names = 1, header=T)

gw.est = as.vector(gw.est)
gw.est = t(gw.est)
names(gw.est) = colnames(gw.est)

wb.web = read.csv("DV_foodwebs_2018/wb_web.csv", row.names = 1, header=T)
wb.est = read.csv("DV_foodwebs_2018/wb_est.csv", row.names = 1, header=T)

wb.est = as.vector(wb.est)
wb.est = t(wb.est)
names(wb.est) = colnames(wb.est)

## GW WEB
plotweb(gw.web, text.rot=45, adj.high=0, adj.low=1, low.spacing = 0.02, 
arrow="down.center", low.abun=gw.est, abuns.type="additional", labsize=1.2)
title(main="Grafham Water")

## WB WEB
plotweb(wb.web, text.rot=45, adj.high=0, adj.low=1, low.spacing = 0.03, 
arrow="down.center", low.abun=wb.est, abuns.type="additional", labsize=1.2)
title(main="Wroxham Broad")

## WRITING LOG FILE FOR R VERSION AND SESSION INFO ##

sink("log_bipartite-session.txt")
sessionInfo()
citation()
sink()


```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```