---
title: "Chapter3-Dv-metabarcoding"
output:
  pdf_document: default
  html_document: default
---

```{r knitr setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r environment setup, include=F}
## R(v3.5.1 - Win)
#.libPaths(c("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win", .libPaths("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win")))

## R(v3.5.1 - OSX)
#.libPaths(c("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx", .libPaths("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx")))

## R(v3.5.1 - Ubuntu)
.libPaths(c("/home/mb/Dropbox/PhD Hull/PhD docs/Thesis/R_stats/rlib3.5_unix", .libPaths("/home/mb/Dropbox/PhD\ Hull/PhD\ docs/Thesis/R_stats/rlib3.5_unix")))
```

```{r directory setup, include=FALSE}
dir()
rm(list=ls())
ls()
```

## Setting working directory and packages

The analysis was done in R (R Core Team, 2018) using the environment and packages as specified below, and in `log_session_packages.txt`.

```{r packages loading, include=F}
pack.list = c("aod","bipartite","boot","cowplot","devtools","dplyr","EcoSimR","ggplot2","ggsignif","gplots","grid","gridExtra","iNEXT","lattice","lme4","lmtest","MASS","mgcv","msa","network","plyr","reshape2","spaa","stringi","tidyr","UpSetR","vegan","zoo")
new.packages = pack.list[!(pack.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages, repos="https://cran.rstudio.com")
lapply(pack.list, require, character.only=T)
```

```{r session}
sessionInfo()
```

The data used in this analysis were generated across 2 separate sequencing runs; one containing gut contentns and eDNA samples collected in May 2017, the second with samples collected in October 2017. Despite the date of the sequencing, all data frames and variable will be named after the month of collection for practicality.

```{r input_files}
infile_may17 = read.csv("data/CO1Aug17-merge-forwonly-nonchimera-c97-cov2_id97-by-taxonomy-readcounts.blast.csv", header=T)
infile_oct17 = read.csv("data/CO1DvAug18-merged-forwonly_nonchimera-c97-cov2-id97_by-taxonomy-readcounts.csv", header=T)
```

Removing the taxonomy column from each input files and saving them into separate files.

```{r removing_taxonomy}
## Custom function to format the taxonomy column

format_taxonomy <- function(input_data){
  for (t in as.data.frame(input_data)){
    n = gsub("\\['|\\']", "", t)
    i = gsub("\\'", "", n)
    taxa.split = data.frame(cbind(i))
  }
  return(taxa.split)
}

# First input file
taxa.may17 = infile_may17$taxomomy
infile_may17 = infile_may17[,!(colnames(infile_may17)%in% 'taxomomy')]

taxa.split.may = format_taxonomy(taxa.may17)
taxa.split.may = data.frame(taxa.split.may)

# Second input file
taxa.oct17 = infile_oct17$taxomomy
infile_oct17 = infile_oct17[,!(colnames(infile_oct17)%in% 'taxomomy')]

taxa.split.oct = format_taxonomy(taxa.oct17)
taxa.split.oct = data.frame(taxa.split.oct)
```

```{r taxa_output, include=TRUE}
# Writing taxonomy outputs
write.csv(taxa.split.may, "R_outputs/CO1DvMay17_taxonomy.csv")
write.csv(taxa.split.oct, "R_outputs/CO1DvOct17_taxonomy.csv")
```

Each initial input file was generated as the result of two separate taxonomic assignment operations. First DNA centroids were assigned against reference database, the DNA reads that resulted unassigned were then assigned against the NCBI nucleotide (nt) database (available at: `ftp://ftp.ncbi.nlm.nih.gov/blast/db/` - updated to early September 2018).
Each operation generated two separate files which were then merged. The following step is to remove potential duplicate taxa records that would prevent R from definying the row names with the taxa names. Because we removed the last column with the taxonomy information, we can use the whole frame; otherwise we would have needed to avoid the last column in each frame.

```{r removing duplicates}
dv.may17 = as.data.frame(infile_may17 %>% group_by(X.OTU.ID) %>% summarise_all(list(sum)))
dv.oct17 = as.data.frame(infile_oct17 %>% group_by(X.OTU.ID) %>% summarise_all(list(sum)))
```

Considering the sequencing runs included different experiments, we exclude the columns containing headers belonging to other experiments. In this case `CH`,`Dv5`,`Ha`, and `pl4`.

```{r extracting_Dv-samples}
rownames(dv.may17) = dv.may17$X.OTU.ID
rownames(dv.oct17) = dv.oct17$X.OTU.ID

dv.may17 = dv.may17[,!(colnames(dv.may17)%in% 'X.OTU.ID')]
dv.oct17 = dv.oct17[,!(colnames(dv.oct17)%in% 'X.OTU.ID')]

to_remove1 = grep("Ha|CH|Dv5", colnames(dv.may17))
dv.may17 = dv.may17[,-to_remove1]

to_remove2 = grep("pl4|CH", colnames(dv.oct17))
dv.oct17 = dv.oct17[,-to_remove2]
```

# 1. Initial Quality Control

To calculate sample read depth for both input files, in order to quality check the overall sequencing, first all samples that didn't sequence at all (sample N read equal to 0) need to be removed. Then the mean and standard deviation for each data set can be calculated, and quality check of the samples read depth applied. 

```{r initial read depth, include=TRUE}
dv.may17 = subset(dv.may17, select = colSums(dv.may17) > 0)
dv.oct17 = subset(dv.oct17, select = colSums(dv.oct17) > 0)

may.read.depth = as.data.frame(dv.may17)
oct.read.depth = as.data.frame(dv.oct17)

may.read.depth$time = "May"
oct.read.depth$time = "Oct"

# Sum of DNA reads by samples to plot the sample read depth

may.read.depth = melt(may.read.depth)
may.read.depth = may.read.depth %>% group_by(time, variable) %>% summarise_all(funs(sum))

oct.read.depth = melt(oct.read.depth)
oct.read.depth = oct.read.depth %>% group_by(time, variable) %>% summarise_all(funs(sum))

# Joining both frames, then plotting

full.read.depth = rbind(may.read.depth, oct.read.depth)

ggplot(full.read.depth, aes(time, value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color=c("gray30"), alpha=0.5) +
  theme_bw() + labs(x="Data sets", y="Sample read depth (N reads)") +
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold"))
```

```{r mean read depth + se}
# May
mean(may.read.depth$value)
sd(may.read.depth$value)/sqrt(length(may.read.depth$value))

# Oct
mean(oct.read.depth$value)
sd(oct.read.depth$value)/sqrt(length(oct.read.depth$value))
```

The sample N reads appear to be highly dispersed, this causes the mean-sd values to result negative. After checking for normality and log-transforming the data sets, the read depth threshold could be applied.
```{r}
quantile(colSums(dv.may17))
quantile(colSums(dv.oct17))
```


```{r mean and sd calculation}
mean(colSums(dv.may17))-sd(colSums(dv.may17))
mean(colSums(dv.oct17))-sd(colSums(dv.oct17))

shapiro.test(colSums(dv.may17))
shapiro.test(colSums(dv.oct17))

may.depth.thre = 500
oct.depth.thre = 500
```

The filter coverage for all samples that are equal and greater than 100 reads coverage per samples were retained. The survival of the N samples was quite high in both data sets, with respectivelty 192 (88.9%) and 190 (95.9%) in May and October data sets.

```{r test_coverage, include=TRUE, echo=FALSE}
# Applying the threshold
reads.may17 = subset(dv.may17, select = colSums(dv.may17) >= may.depth.thre)
reads.oct17 = subset(dv.oct17, select = colSums(dv.oct17) >= oct.depth.thre)

# Checking N samples retained
count(colSums(dv.may17) <= may.depth.thre)[1,2]
count(colSums(dv.oct17) <= oct.depth.thre)[1,2]
count(colSums(dv.may17) <= may.depth.thre)[1,2]/(count(colSums(dv.may17) > 0)[,2])
count(colSums(dv.oct17) <= oct.depth.thre)[1,2]/(count(colSums(dv.oct17) > 0)[,2])
```

Unwanted taxa will now need to be removed before the contamination control.

```{r cleaning_frames}
# Cleaning empty taxa
reads.may17 = subset(reads.may17, subset = rowSums(dv.may17) > 0)
reads.oct17 = subset(reads.oct17, subset = rowSums(dv.oct17) > 0)

disc.may = c("Abramis_brama","Anser","Aphanomyces","Arthropoda_environmental_sample","Batrachospermum_gelatinosum","Chromulinales","Cochliopodium","Cochliopodium_actinophorum","Cyclostephanos_dubius","Cyclostephanos_invisitatus","Cyclostephanos_sp._WTC16","Didymella_pinodes","Dinophyceae","Encyonema","Eukaryota","Fistulifera_solaris","Fusarium","Gomphonema_parvulum","Gymnocephalus","Hominidae","Homo_sapiens","Ilyonectria_destructans","invertebrate_environmental_sample","Lagenidium","Leptolegnia_sp._BOLD:AAX5717","Mallomonas_sp.","Melosira_ambiqua","Metazoa","Mus_musculus","Oncorhynchus_mykiss","Oomycetes","Paraphysomonas_sp.","Penicillium_sclerotiorum","Phytophthora","Pinnularia_neomajor","Pseudorasbora_parva","Pythium","Pythium_marsipium","Saprolegnia_delica","Saprolegniaceae","Sellaphora_obesa","Skeletonema_potamos","Sordariomycetes","Stephanodiscus","Stephanodiscus_cf._minutulus","Sus_scrofa","Tetracladium","Tetradesmus_obliquus","Thalassiosira_pseudonana","Urocentrum_turbo","unassigned","Yamagishiella_unicocca","Woloszynskia","zooplankton_environmental_sample")

clean.may = reads.may17[!(rownames(reads.may17) %in% disc.may),]

disc.oct = c("Abramis_brama","Achlya","Actinopteri","Albugo_candida","Anatidae","Anser_anser","Aphanomyces","Aploneura_lentisci","Apoikiospumella_mondseeiensis","Ascomycota","Aves","Batrachospermaceae","Batrachospermum_gelatinosum","Bodo_saltans","Branta_canadensis","Chlorella_sp._ArM0029B","Chroicocephalus_ridibundus","Chromulinales","Chrysochromulina","Chrysochromulina_parva","Chrysophyceae_sp.","Cochliopodium","Colletotrichum","Cordylophora_sp._IM1","Cordylophora_sp._NFR3","Cristatella_mucedo","Cyclostephanos_dubius","Cyclostephanos_invisitatus","Cyclostephanos_sp._WTC16","Cygnus_olor","Cylindrodendrum_hubeiense","Didymella_pinodes","Didymium_minus","Dinobryon","Dinobryon_pediforme","Dinophyceae","Emericellopsis","Eukaryota","Fistulifera_solaris","Fulica_atra","Fungi","Fusarium","Gomphonema_parvulum","Gymnocephalus_cernua","Heterolepidoderma_macrops","Homo_sapiens","Hypocreales","Ilyonectria_destructans","invertebrate_environmental_sample","Mallomonas_sp.","Melampsora","Melosira_ambiqua","Mesostigma_viride","Metazoa","Microcystis","Microcystis_aeruginosa","Mucorales","Myodes_glareolus","Nitzschia_palea","Nusuttodinium","Oncorhynchus_mykiss","Oomycetes","Ostreopsis","Paraphysomonas_sp.","Pedospumella","Penicillium","Percidae","Phytophthora","Phytophthora_gallica","Phytophthora_plurivora","Plumatella_fungosa","Podiceps_cristatus","Prunella_modularis","Pseudomuriella","Pseudoperonospora_humuli","Pycnococcaceae","Pythiaceae","Pythium","Pythium_marsipium","Pythium_phragmitis","Pythium_sp._JN-1","Ripella","Rutilus_rutilus","Salmo_trutta","Saprolegnia_delica","Scardinius","Segregatospumella_dracosaxi","Sellaphora_bacillum","Sellaphora_lanceolata","Sellaphora_obesa","Sellaphora_pupula","Sheathia_arcuata","Skeletonema","Spumella_sp.","Spumella_sp._LG-2014b","Stephanodiscus","Streptophyta","Tetracladium","Tetracladium_setigerum","Tetradesmus_obliquus","Thalassiosira_pseudonana","Thelonectria_discophora","Thelonectria_veuillotiana","Turdus_merula","unassigned","uncultured_Jaminaea","uncultured_Philodina","uncultured_Pythium","Vermamoeba_vermiformis","Verticillium","Verticillium_dahliae","zooplankton_environmental_sample","Zymoseptoria_tritici")

clean.oct = reads.oct17[!(rownames(reads.oct17) %in% disc.oct),]
```

The contamination control is based maninly on the `POSITIVE` samples. Those included single-specie tissue DNA of _Osmia bicornis_ (Linnaeus, 1758); and the proportion of other taxa detected in those single-specie samples will be the base on which the contamination threshold will be calculated.

```{r contamination control}
pos.may = clean.may[,grep("POS", colnames(clean.may))]
pos.oct = clean.oct[,grep("POS", colnames(clean.oct))]

colSums(pos.may)
colSums(pos.oct)

# We remove all taxa equal to 0

pos.may = subset(pos.may, subset = rowSums(pos.may) > 0)
pos.oct = subset(pos.oct, subset = rowSums(pos.oct) > 0)

# We decided to ignore one of the positive samples from May 2017 as it contained only 22 reads, of which none belonged to positive taxa _Osmia bicornis_

pos.may = subset(pos.may, select = colSums(pos.may) > 22)

colSums(pos.may["Osmia_bicornis",])
colSums(pos.oct["Osmia_bicornis",])

colSums(pos.may)-colSums(pos.may[c("Osmia_bicornis","Hymenoptera"),])
colSums(pos.oct)-colSums(pos.oct[c("Osmia_bicornis","Hymenoptera"),])

(colSums(pos.may)-colSums(pos.may[c("Osmia_bicornis","Hymenoptera"),]))/colSums(pos.may)
(colSums(pos.oct)-colSums(pos.oct[c("Osmia_bicornis","Hymenoptera"),]))/colSums(pos.oct)
```

Three of the positive samples in the October data set contained high amount of contamination from _Crangonyx floridanus_, and _Dikerogammarus villosus_ DNA. 
Specifically 648 reads (46.3%) belonging to _C. floridanus_ were present in a single positive sample; and 2106 reads (68.3%) and 2681 reads (71.6%) belonging to _D. villosus_ in two separate positive samples. In all these cases the proportion of reads was going to be too high to be used as threshold. The decision taken was then to exclude those three samples before calculating the contamination threshold for the October data set.

```{r high levels taxa}
pos.oct["Crangonyx_floridanus",]
pos.oct["Crangonyx_floridanus",]/colSums(pos.oct)
pos.oct["Dikerogammarus_villosus",]
pos.oct["Dikerogammarus_villosus",]/colSums(pos.oct)

# Removing the high contaminated positive samples
pos.oct = pos.oct[,-c(1:3)]
```

Based on the levels of contamination in the Positive samples, the threshold levels were decided to be respectively 0.03% (May data set) and 0.5% (October data set). In May data set we still detect _D. villosus_ contamination.

```{r threshold May}
# Transposing the data frames and copying them to carry them over as N reads and ratios.
clean.may = as.data.frame(t(clean.may))

filter.may = as.data.frame(clean.may)
filter.may = filter.may[,]/rowSums(filter.may)
filter.may[is.na(filter.may)] <- 0

filter.may[][filter.may <= 0.0003] <- 0
clean.may[][clean.may[,]/rowSums(clean.may) <= 0.0003] <- 0
```

```{r threshold October}
clean.oct = as.data.frame(t(clean.oct))

filter.oct = as.data.frame(clean.oct)
filter.oct = filter.oct[,]/rowSums(filter.oct)
filter.oct[is.na(filter.oct)] <- 0

filter.oct[][filter.oct <= 0.005] <- 0
clean.oct[][clean.oct[,]/rowSums(clean.oct) <= 0.005] <- 0

# Removing empty taxa

clean.may = subset(clean.may, select = colSums(clean.may) > 0)
filter.may = subset(filter.may, select = colSums(filter.may) > 0)

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)
```

# 2. Checking for remaining known contaminations

Checking for remaining contamination, _Harmonia axyridis_ is still present in one samples of May data set (RB501) with 7 reads (3.9%). Similarly, _Osmia bicornis_ is still present in few samples; however excluding the POSITIVE samples, the remaining contamination is in very low amount. Despite the contamination filter threshold didn't manage to remove these issues, the DNA read count is low enough to allow removal without major changes in the final output. Similar situation happened for the Primer contamination labeled `cont.13` which managed to pass the initial QC threshold despite having 17 reads (11 belonging to _O. bicornis_ and 6 belonging to _D. villosus_). The read count is low enough to feel confident in not having an influence, however being fairly widespread the decision is to remove all _D. villosus_ detections in the gut contents. The removal will happen further downstream to prevent mis-identifying the predators.

```{r remaining contaminations May}
# Checking in which sample(s) _Osmia bicornis_ contamination is still present
count(filter.may$Osmia_bicornis > 0)
rownames(filter.may)[filter.may$Osmia_bicornis > 0]

# Checking N reads of _O. bicornis_
clean.may$Osmia_bicornis[rownames(clean.may) %in% rownames(filter.may)[filter.may$Osmia_bicornis > 0]]

clean.may = clean.may[,!(colnames(clean.may) %in% "Osmia_bicornis")]
filter.may = filter.may[,!(colnames(filter.may) %in% "Osmia_bicornis")]

# Checking in which sample(s) _Harmonia axyridis_ contamination is still present
count(filter.may$Harmonia_axyridis > 0)
rownames(filter.may)[filter.may$Harmonia_axyridis > 0]

# Checking N reads of _H. axyridis_ 
clean.may$Harmonia_axyridis[rownames(clean.may) %in% rownames(filter.may)[filter.may$Harmonia_axyridis > 0]]

# Removing _H. axyridis_ without need of further filtration
clean.may = clean.may[,!(colnames(clean.may) %in% "Harmonia_axyridis")]
filter.may = filter.may[,!(colnames(filter.may) %in% "Harmonia_axyridis")]
```

In the samples from October, the contamination situation was less problematic. No _Harmonia axyridis_ remained after filter threshold; and _Osmia bicornis_ remained in only one samples (eGW4) with 28 reads (1.88%). In addition it was present contamination from _Porcellio scaber_ in two samples (RB409 and RB412), respectively at 2656 reads (5.6%), and 6234 reads (16.9%). Ultimately Psocoptera DNA was detected as contaminant in one sample (RB411), at 293 reads (0.9%).


```{r remaining contaminations Oct}
# Checking in which sample(s) _Osmia bicornis_ contamination is still present
count(filter.oct$Osmia_bicornis > 0)
rownames(filter.oct)[filter.oct$Osmia_bicornis > 0]

# Checking N reads of _O. bicornis_
clean.oct$Osmia_bicornis[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$Osmia_bicornis > 0]]

# Checking in which sample(s) _Psocoptera_ contamination is still present
count(filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0)
rownames(filter.oct)[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0]

# Checking N reads of _Psocoptera_
clean.oct$`Psocoptera_sp._BOLD:AAN8452`[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0]]
filter.oct[filter.oct$`Psocoptera_sp._BOLD:AAN8452` > 0,"Psocoptera_sp._BOLD:AAN8452"]

# Checking in which sample(s) _Porcellio scaber_ contamination is still present
count(filter.oct$Porcellio_scaber > 0)
rownames(filter.oct)[filter.oct$Porcellio_scaber > 0]

# Checking N reads of _P. scaber_
clean.oct$Porcellio_scaber[rownames(clean.oct) %in% rownames(filter.oct)[filter.oct$Porcellio_scaber > 0]]
filter.oct[filter.oct$Porcellio_scaber > 0,"Porcellio_scaber"]

# Removing contamination
clean.oct = clean.oct[,!(colnames(clean.oct) %in% c("Osmia_bicornis","Psocoptera_sp._BOLD:AAN8452","Porcellio_scaber"))]
filter.oct = filter.oct[,!(colnames(filter.oct) %in% c("Osmia_bicornis","Psocoptera_sp._BOLD:AAN8452","Porcellio_scaber"))]

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)

clean.oct = subset(clean.oct, subset = rowSums(clean.oct) > 0)
filter.oct = subset(filter.oct, subset = rowSums(filter.oct) > 0)
```

The DNA extraction blanks (`GW.BL1 & 2`, `WB.BL1 & 2`, `RB.BL1 & 2`) resulted containing quite a high amount of contaminant DNA reads:

* `GW.BL1` = 3399, `GW.BL2` = 1782
* `WB.BL1` = 6338, `WB.BL2` = 863
* `RB.BL1` = 1818, `RB.BL2` = 2002

Of these, between 99.4% and 100% belonged to _Dikerogamamrus villosus_. This is far less than ideal considering those are meant to be dissection and extraction blanks; however the rest of the non-invaded Rockland Broad (RB) gut content samples presented very low _D. villosus_ contamination. In this perspective the decision was made not to act upon these levels of contamination also considering that _D. villosus_ is the main target taxa and overall in high concentration.

```{r extraction blanks}
# cleaning contamination in extraction blank samples
rowSums(clean.oct[grep("BL", rownames(filter.oct)),])
clean.oct[grep("BL", rownames(filter.oct)),"Dikerogammarus_villosus"]/rowSums(clean.oct[grep("BL", rownames(filter.oct)),])

clean.oct[grep("RB", rownames(filter.oct)),"Dikerogammarus_villosus"]

clean.oct = clean.oct[-c(grep("BL", rownames(filter.oct))),]
filter.oct = filter.oct[-c(grep("BL", rownames(filter.oct))),]

clean.oct = subset(clean.oct, select = colSums(clean.oct) > 0)
filter.oct = subset(filter.oct, select = colSums(filter.oct) > 0)

clean.oct = subset(clean.oct, subset = rowSums(clean.oct) > 0)
filter.oct = subset(filter.oct, subset = rowSums(filter.oct) > 0)
```

Positive, Negative and Blank samples need to be removed before continuing.

```{r removing samples}
pnmay = grep("POS|NEG", rownames(filter.may), ignore.case=T)
filter.may = filter.may[-pnmay,]
clean.may = clean.may[-pnmay,]

pnoct = grep("POS|NEG", rownames(filter.oct), ignore.case=T)
filter.oct = filter.oct[-pnoct,]
clean.oct = clean.oct[-pnoct,]
```

Other variables need to be added to the data frames; specifically the sampling season, the gut contents samples ID, the site ID, and the predator ID (eDNA samples were labeled as "water"). Then the frames can be divided between gut contents and eDNA samples.

```{r adding variables}
# Adding time variable
filter.may$time = "May"
clean.may$time = "May"

filter.oct$time = "Oct"
clean.oct$time = "Oct"

# Adding sample ID from rownames
# May
filter.may$sample_id = gsub("[.]", "", rownames(filter.may))
clean.may$sample_id = gsub("[.]", "", rownames(clean.may))

# October
filter.oct$sample_id = gsub("[.]pl[0-9]{1}.{5,6}nc.blast", "", rownames(filter.oct))
clean.oct$sample_id = gsub("[.]pl[0-9]{1}.{5,6}nc.blast", "", rownames(clean.oct))

# Nested gsub to keep only the 2 letter from each sample ID referring to the lake ID
# May
filter.may$site = gsub("BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", filter.may$sample_id))))
clean.may$site = gsub("BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", clean.may$sample_id))))

# October
filter.oct$site = gsub("[.]BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", filter.oct$sample_id))))
clean.oct$site = gsub("[.]BL{1}$", "", gsub("R{1}$", "", gsub("[0-9]{1,3}", "", gsub("e?", "", clean.oct$sample_id))))

# Adding the sample type (eDNA or gut contents)
# May
ednamay = grep("e", filter.may$sample_id)
filter.may$sample_type = "NA"
filter.may[ednamay,"sample_type"] = "eDNA"
filter.may[-ednamay,"sample_type"] = "gut"

ednamay = grep("e", clean.may$sample_id)
clean.may$sample_type = "NA"
clean.may[ednamay,"sample_type"] = "eDNA"
clean.may[-ednamay,"sample_type"] = "gut"

# October
ednaoct = grep("e", filter.oct$sample_id)
filter.oct$sample_type = "NA"
filter.oct[ednaoct, "sample_type"] = "eDNA"
filter.oct[-ednaoct, "sample_type"] = "gut"

ednaoct = grep("e", clean.oct$sample_id)
clean.oct$sample_type = "NA"
clean.oct[ednaoct, "sample_type"] = "eDNA"
clean.oct[-ednaoct, "sample_type"] = "gut"
```

For the `predator_id`, the taxa ID was extracted from the taxa with ratio greater or equal to 90% within each samples, for both May and October. The samples that resulted still as "NA" at the end of this process were manually checked and the taxa ID were added manually.

```{r adding predator}
# Adding predator ID based on the DNA ratio greater than 0.9
# May
ednamay = grep("e", rownames(filter.may))
predmay = which(filter.may[-ednamay,!c(colnames(filter.may) %in% c("time","site","sample_id","sample_type"))] >= 0.9, arr.ind=T)

filter.may$predator_id = "NA"
filter.may[ednamay,"predator_id"] = "water"
filter.may[rownames(predmay),"predator_id"] = gsub("[.]?[0-9]{1,3}$", "", colnames(filter.may[rownames(predmay), as.vector(predmay[,2])]))

# October
ednaoct = grep("e", rownames(filter.oct))
predoct = which(filter.oct[-ednaoct,1:74] >= 0.9, arr.ind=T)

filter.oct$predator_id = "NA"
filter.oct[ednaoct,"predator_id"] = "water"
filter.oct[rownames(predoct),"predator_id"] = gsub("[.]?[0-9]{1,3}$", "", colnames(filter.oct[rownames(predoct),as.vector(predoct[,2])]))

# Checking for missing predators
# May
count(filter.may$predator_id == "NA")
miss.may = gsub("[.]", "", rownames(filter.may[filter.may$predator_id == "NA",]))
miss.may

filter.may$predator_id[filter.may$sample_id %in% "WB507"] = "Dikerogammarus_villosus"
filter.may$predator_id[filter.may$sample_id %in% c("RB304","RB303")] = "Corophium_multisetosum"
filter.may$predator_id[filter.may$sample_id %in% c(miss.may[!miss.may %in% c("WB507","RB304","RB303")])] = "Gammarus_zaddachi"
clean.may$predator_id = filter.may$predator_id

# October
count(filter.oct$predator_id == "NA")
miss.oct = rownames(filter.oct[filter.oct$predator_id == "NA",])
miss.oct

filter.oct$predator_id[filter.oct$sample_id %in% c("GW103","GW108","GW401","GW601","WB627")] = "Dikerogammarus_villosus"
filter.oct$predator_id[filter.oct$sample_id %in% c("RB301","RB302","RB303","RB304","RB305","RB606","RB613","RB616")] = "Gammarus_zaddachi"
filter.oct$predator_id[filter.oct$sample_id %in% c("RB403","RB412")] = "Crangonyx_pseudogracilis"
clean.oct$predator_id = filter.oct$predator_id
```

Now that all variable have been added, the ratio and DNA reads of predators can be set to be equal to 0. This will allow further downstream analysis to calulate accurate indexes of prey diversity and diet breadth without the risk of overinflating.

```{r removing predator ratio}
# Removal of predator ratios and DNA reads
# May
m.row = rownames(filter.may[filter.may$predator_id != "water", c("site","sample_id","time","sample_type","predator_id")])
m.col = filter.may[filter.may$predator_id != "water", "predator_id"]

filter.may[m.row,unique(m.col)]

for (i in m.row){
  pred_id = filter.may[i,"predator_id"]
  pred_read = clean.may[i, "predator_id"]
  filter.may[i,pred_id] <- 0
  clean.may[i,pred_read] <- 0
}

# Confirming the function worked
filter.may[m.row,unique(m.col)]

# Oct
o.row = rownames(filter.oct[filter.oct$predator_id != "water", c("site","sample_id","time","sample_type","predator_id")])
o.col = filter.oct[filter.oct$predator_id != "water", "predator_id"]

filter.oct[o.row,unique(o.col)]

for (i in o.row){
  pred_id = filter.oct[i,"predator_id"]
  pred_read = clean.oct[i, "predator_id"]
  filter.oct[i,pred_id] <- 0
  clean.oct[i,pred_read] <- 0
}

# Confirming the function worked
filter.oct[o.row,unique(o.col)]
```

Before preocessing the gut contents info, we calculate predator to prey DNA ratio.

```{r prey-predator ratio}
mean(c(rowSums(filter.may[filter.may$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100))
sd(c(rowSums(filter.may[filter.may$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100))/sqrt(length(c(rowSums(filter.may[filter.may$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Dikerogammarus_villosus", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100)))

mean(c(rowSums(filter.may[filter.may$predator_id == "Gammarus_zaddachi", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Gammarus_zaddachi", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100))
sd(c(rowSums(filter.may[filter.may$predator_id == "Gammarus_zaddachi", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Gammarus_zaddachi", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100))/sqrt(length(c(rowSums(filter.may[filter.may$predator_id == "Gammarus_zaddachi", !c(colnames(filter.may) %in% c("site","sample_id","time","sample_type","predator_id"))])*100,rowSums(filter.oct[filter.oct$predator_id == "Gammarus_zaddachi", !c(colnames(filter.oct) %in% c("site","sample_id","time","sample_type","predator_id"))])*100)))
```

Now that predators DNA ratio have been removed, the focus can move towards the prey id. The current data are going to be saved into ".csv" files before continuing.

```{r saving files}
write.csv(clean.may, "R_outputs/May_reads_filtered.csv")
write.csv(filter.may, "R_outputs/May_ratio_filtered.csv")
write.csv(clean.oct, "R_outputs/Oct_reads_filtered.csv")
write.csv(filter.oct, "R_outputs/Oct_ratio_filtered.csv")
```

Dividing gut contents and eDNA samples into separate dataframes.

```{r dividing samples}
clean.guts = full_join(clean.may[clean.may$sample_type == "gut",], clean.oct[clean.oct$sample_type == "gut",])
clean.guts[is.na(clean.guts)] <- 0

clean.edna = full_join(clean.may[clean.may$sample_type == "eDNA",], clean.oct[clean.oct$sample_type == "eDNA",])
clean.edna[is.na(clean.edna)] <- 0
```

```{r removing empty taxa}
# Removing D.villosus contaminations from Rockland Broad (max read = 36)
clean.edna[grep("RB", clean.edna$site), grep("Dikerogammarus", colnames(clean.edna))] <- 0

# Extracting metadata
info.guts = clean.guts[,c(colnames(clean.guts) %in% c("time","site","sample_id","sample_type","predator_id"))]
info.edna = clean.edna[,c(colnames(clean.edna) %in% c("time","site","sample_id","sample_type","predator_id"))]

# Saving gut contents and eDNA files
write.csv(clean.guts, "R_outputs/gut_contents_reads.csv")
write.csv(clean.edna, "R_outputs/eDNA_reads.csv")

# Cleaning empty taxa
# gut content samples
clean.guts = subset(clean.guts[,!c(colnames(clean.guts) %in% colnames(info.guts))], select=colSums(clean.guts[,!c(colnames(clean.guts) %in% colnames(info.guts))]) > 0)
# eDNA samples
clean.edna = subset(clean.edna[,!c(colnames(clean.edna) %in% colnames(info.edna))], select=colSums(clean.edna[,!c(colnames(clean.edna) %in% colnames(info.edna))]) > 0)
```

# 3. Detection of prey DNA in gut contents

The predator DNA reads now are equal to 0 across all gut content samples. This allows for calculating the prey DNA ratio within each gut content sample, that will allow also for a niche overlap analysis.
the prey species detected included also both _Crangonyx pseudogracilis_ and _C. floridanus_. Becuase _C.floridanus_ has been detected in the UK, but hasn't been confirmed in Rockland Broad, the decision was taken to be conservative and pull both Crangonyx species up to the genus level summing their DNA reads.
Furthermore the taxa `Isopoda`, that should have been assigned to _Asellus aquaticus_, can't really be used for our purposes because of our detected contamination from _Porcellio scaber_. The 1217 read belonging to Isopoda will need to be removed from the gut contents dataset. 
_Dikerogammarus villosus_ resulted to be fairly spread as contamination across the dataset. The decision was made to remove it from the gut contents data set, after the predator ID was extracted.

```{r cleaning gut contents data frame}
# Removing high rank taxa and Dikerogammarus villosus
clean.guts = clean.guts[,!c(colnames(clean.guts) %in% c("Isopoda","Diptera","Dikerogammarus_villosus","Cyclopidae"))]

# converting reads into ratio
ratio.guts = clean.guts[,]/rowSums(clean.guts)
ratio.guts[is.na(ratio.guts)] <- 0

ratio.guts = subset(ratio.guts, rowSums(ratio.guts) > 0)

# Using row names to subset the info associated with the gut contents
rownames(ratio.guts)
info.guts = info.guts[rownames(ratio.guts),]
```

The miss-identification of few predators is reflected in the results of the sequencing on the gut contents. So far all predators were included to allow for accurate quality controls; however before continuing into more semi-quantitative analysis of overlap and diversity indexes those mis-identified predators will need to be removed. 11 samples corresponding to _Corophium multisetosum_ (N=2), _Gammarus tigrinus_ (N=2), and _Crangonyx pseudogracilis_ (N=7) will need to be removed from the dataset as they won't provide accurate comparisons with the rest of the dataset.

```{r prey ratio in guts}
ratio.guts = cbind(ratio.guts, info.guts)
ratio.guts = ratio.guts[ratio.guts$predator_id %in% c("Dikerogammarus_villosus","Gammarus_zaddachi"),]
melt.guts = melt(ratio.guts)
```

```{r initial plot}
ggplot(melt.guts, aes(time, value)) +
  geom_jitter(aes(color=predator_id)) +
  facet_grid (.~site) + theme_bw() +
  ylab("prey DNA ratio") + xlab("sampling month")
```


```{r prey detection rarefaction curves}
# removing `sample_type` and `sample_id` variables and transforming detections into abundances
abund.guts = as.data.frame(ratio.guts)
abund.guts[,!c(colnames(abund.guts) %in% c("time","site","predator_id","sample_id","sample_type"))][abund.guts[,!c(colnames(abund.guts) %in% c("time","site","predator_id","sample_id","sample_type"))] > 0] <- 1

# rarefaction curves
abund.guts.site = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("sample_id","sample_type"))] %>% group_by(time, site, predator_id) %>% summarise_all(funs(sum)))

# replacing predator id with initials
abund.guts.site[grep("Gammarus_zaddachi", abund.guts.site$predator_id),"predator_id"] <- "G.zad"
abund.guts.site[grep("Dikerogammarus_villosus", abund.guts.site$predator_id),"predator_id"] <- "D.vil"

abund.guts[grep("Gammarus_zaddachi", abund.guts$predator_id),"predator_id"] <- "G.zad"
abund.guts[grep("Dikerogammarus_villosus", abund.guts$predator_id),"predator_id"] <- "D.vil"

# setting rownames and removing extra variables
rownames(abund.guts.site) = paste(abund.guts.site$predator_id, abund.guts.site$site, abund.guts.site$time, sep="_")
inext.guts = t(abund.guts.site[,!c(colnames(abund.guts.site) %in% c("site","time","predator_id"))])

inext.guts = iNEXT(inext.guts, q=c(0,1), datatype = "abundance", se=T, endpoint=200)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(inext.guts) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Species level rarefaction curves of preys in gut contents")
```

```{r PERMANOVA gut contents}
# comparing only D.vil
adonis.dvil = abund.guts[abund.guts$predator_id == "D.vil",]

fact_site.dvil = matrix(c(adonis.dvil$site), ncol=1)
fact_time.dvil = matrix(c(adonis.dvil$time), ncol=1)

adonis.dvil = adonis.dvil[,!c(colnames(adonis.dvil) %in% c("time","site","sample_type","predator_id","sample_id"))]

adonis.dvil = vegdist(as.matrix(adonis.dvil), method = "euclidean")
adonis(adonis.dvil~fact_site.dvil*fact_time.dvil, permutations = 9999, method = "euclidean")

# comparing D.vil and G.zad
adonis.guts = abund.guts[,!c(colnames(abund.guts) %in% c("time","site","sample_type","predator_id","sample_id"))]

fact_time.guts = matrix(c(abund.guts$time), ncol=1)
fact_site.guts = matrix(c(abund.guts$site), ncol=1)
fact_predator.guts = matrix(c(abund.guts$predator_id), ncol=1)

adonis.guts = vegdist(as.matrix(adonis.guts), method = "euclidean")
adonis(adonis.guts~fact_predator.guts*fact_site.guts*fact_time.guts, permutations = 9999, method="euclidean")
```

```{r NMDS on D.vil gut contents}
meta.dv.guts = metaMDS(adonis.dvil, trymax = 200, distance = "euclidean")
```

```{r NMDS of gut contents both D.vil and G.zad}
meta.guts = metaMDS(adonis.guts, trymax=200, distance="euclidean")
```

```{r nmds of d.villosus guts}
# full symbols for May, hollow symbols for October
png("R_images/guts_Dvil_time_NMDS_euclidean.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.dv.guts, type="n", main="D.villosus gut contents NMDS by time", xlim=c(-1,2), ylim=c(-1,1))
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="WB" & fact_time.dvil[,1]=="May"),col="black",pch=20)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="WB" & fact_time.dvil[,1]=="Oct"),col="black",pch=1)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="GW" & fact_time.dvil[,1]=="May"),col="blue",pch=15)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="GW" & fact_time.dvil[,1]=="Oct"),col="blue",pch=22)
ordispider(meta.dv.guts,group=fact_site.dvil[,1],col=c("blue","green"),label=F)
ordiellipse(meta.dv.guts,group=fact_time.dvil[,1],kind="ehull",lty=c(1,5),label=T,lwd=1)
text(-0.5,-1.1,label=paste("Stress:", round(meta.dv.guts$stress, 6)))
dev.off()
```

```{r nmds of d.villosus guts by site}
# full symbols for May, hollow symbols for October
png("R_images/guts_Dvil_only_NMDS_euclidean.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.dv.guts, type="n", main="D.villosus gut contents NMDS by site", xlim=c(-1,2), ylim=c(-1,1))
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="WB" & fact_time.dvil[,1]=="May"),col="black",pch=20)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="WB" & fact_time.dvil[,1]=="Oct"),col="black",pch=1)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="GW" & fact_time.dvil[,1]=="May"),col="blue",pch=15)
points(meta.dv.guts, select=which(fact_site.dvil[,1]=="GW" & fact_time.dvil[,1]=="Oct"),col="blue",pch=22)
ordispider(meta.dv.guts,group=fact_site.dvil[,1],col=c("blue","green"),label=F)
ordiellipse(meta.dv.guts,group=fact_site.dvil[,1],kind="ehull",lty=c(1,5),label=T,lwd=1)
text(-0.5,-1.1,label=paste("Stress:", round(meta.dv.guts$stress, 6)))
dev.off()
```

```{r nmds of d.villosus and g.zaddachi guts}
# full symbols for May, hollow symbols for October
png("R_images/guts_predator_NMDS_euclidean.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.guts, type="n", main="D.villosus and G.zaddachi gut contents NMDS", xlim=c(-2,1.5), ylim=c(-1.5,1.2))
points(meta.guts, select=which(fact_site.guts[,1]=="WB" & fact_time.guts[,1]=="May"),col="black",pch=20)
points(meta.guts, select=which(fact_site.guts[,1]=="WB" & fact_time.guts[,1]=="Oct"),col="black",pch=1)
points(meta.guts, select=which(fact_site.guts[,1]=="GW" & fact_time.guts[,1]=="May"),col="blue",pch=15)
points(meta.guts, select=which(fact_site.guts[,1]=="GW" & fact_time.guts[,1]=="Oct"),col="blue",pch=22)
points(meta.guts, select=which(fact_site.guts[,1]=="RB" & fact_time.guts[,1]=="May"),col="red",pch=17)
points(meta.guts, select=which(fact_site.guts[,1]=="RB" & fact_time.guts[,1]=="Oct"),col="red",pch=24)
ordispider(meta.guts,group=fact_predator.guts[,1],col=c("blue","orange"),label=T)
ordiellipse(meta.guts,group=fact_predator.guts[,1],kind="ehull",lty=c(1,5),label=F,lwd=1)
text(-1.5,-1.2,label=paste0("Stress: ",round(meta.guts$stress, 6)))
dev.off()
```

# 4. Niche/Diet breadth

After analysisng the detection of prey species in the gut contents, and after having analysed the communities compositions from eDNA and kick samples. The niche breadth of _D. villosus_ across the 2 invaded sites can be analysed. Further, the comparison of niche overlap between _D. villosus_ and _G. zaddachi_ can be calculated.

```{r niche breadth of amphipods across sites}
# Summing
breadth.dvgz = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("time","sample_type","sample_id"))] %>% group_by(predator_id, site) %>% summarise_all(funs(sum)))

rownames(breadth.dvgz) = paste(breadth.dvgz$predator_id, breadth.dvgz$site, sep="_")

breadth.dvgz = t(breadth.dvgz[,!c(colnames(breadth.dvgz) %in% c("predator_id","site"))])

niche.dvgz = breadth.dvgz

breadth.dvgz[,1] = breadth.dvgz[,1]/102 
breadth.dvgz[,2] = breadth.dvgz[,2]/114
breadth.dvgz[,3] = breadth.dvgz[,3]/107

niche.width(breadth.dvgz, method="levins")
niche.width(breadth.dvgz, method="shannon")

# D.villosus GW
(niche.width(breadth.dvgz, method="levins")[1]-1)/(nrow(breadth.dvgz)-1)
# D.villosus WB
(niche.width(breadth.dvgz, method="levins")[2]-1)/(nrow(breadth.dvgz)-1)
# G.zaddachi
(niche.width(breadth.dvgz, method="levins")[3]-1)/(nrow(breadth.dvgz)-1)

niche.overlap(breadth.dvgz, method="levins")
niche.overlap(breadth.dvgz, method="pianka")
```

```{r niche breadth of amphipods species (no sites)}
# summing prey DNA reads by predator ID
breadth.dv = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("time","sample_type","site","sample_id"))] %>% group_by(predator_id) %>% summarise_all(funs(sum)))

rownames(breadth.dv) = breadth.dv$predator_id

breadth.dv = t(breadth.dv[,!c(colnames(breadth.dv) %in% c("predator_id"))])

breadth.dv[,1] = breadth.dv[,1]/216
breadth.dv[,2] = breadth.dv[,2]/107

niche.width(breadth.dv, method="levins")
niche.width(breadth.dv, method="shannon")

(niche.width(breadth.dv, method="levins")[1]-1)/(nrow(breadth.dv)-1)
(niche.width(breadth.dv, method="levins")[2]-1)/(nrow(breadth.dv)-1)

niche.overlap(breadth.dv, method="levins")
niche.overlap(breadth.dv, method="pianka")
```

# 5. Food web bipartite network

For the bipartite network the gut contents data frame with the detections of preys can now be used. The gut contents data first needs to be summarised by site and season, then the data frame can be transposed to have a matrix with the prey along the rows and the predators along the columns. The prey taxa names (bottom level) will also be simplified from the extended binomial names only to allow a better reading in the bipartite plot.

```{r food web network plus bipartite}
# Summing prey DNA reads by predator ID, site and time.
food.web = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("sample_type","sample_id"))] %>% group_by(site, predator_id, time) %>% summarise_all(funs(sum)))

# Renaming for pubblication
food.web[food.web$predator_id == "D.vil", "predator_id"] <- "Dv"
food.web[food.web$predator_id == "G.zad", "predator_id"] <- "Gz"

rownames(food.web) = paste(food.web$predator_id, food.web$site, food.web$time)

# Removing metadata column and transposing to have predators in columns and prey in rows.
food.web = food.web[,!c(colnames(food.web) %in% c("site","predator_id","time"))]
food.web = as.data.frame(t(food.web))
food.web = subset(food.web, subset=rowSums(food.web) > 0)

# Renaming the preys taxa in a more legible format for the plot (e.g. from extended binomial "Asellus aquaticus" to "A.aquaticus")
rownames(food.web) = gsub("[a-z]{3,12}_",".", rownames(food.web))
```

Setting up the order of the top and bottom levels in the bipartite plot, this will allow to colour code the taxa. The taxa in this plot were organised by higher rank (e.g. Diptera, Amphipoda, Oligochaeta, etc.). While the order of the taxa in the plot is organised according to the interactions, in the legend the order was organised by high rank to make it more accessible and easy to read.

```{r plot order}
seq.high=c("Dv GW May","Dv GW Oct","Dv WB May","Dv WB Oct","Gz RB May","Gz RB Oct")

seq.low=c("C.luridus","C.laricomalis","T.tubifex","A.chlorotica","P.acuta","D.polymorpha","G.tigrinus","Cricotopus","C.brevilabris","Daphnia","P.pediculus","G.zaddachi","P.moldaviensis","B.vejdovskyanum","L.hoffmeisteri","C.floridanus","C.pseudogracilis","C.multisetosum","C.sp._BOLD:AAI4299","Hydra","H.oligactis","C.diastrophus","S.lacustris","S.sthenum","A.sieboldi","E.dilatata","L.phaeopa","Chironomidae","T.medius","S.crystallina")

col.network=c("yellow","yellow","tan4","tan4","darkorchid3","darkorchid3","red","yellow","cyan3","cyan3","cyan3","red","tan4","tan4","tan4","red","red","red","yellow","darkcyan","darkcyan","tan4","tan4","grey48","darkgreen","darkgreen","khaki3","yellow","yellow","cyan3")
```

```{r legend order}
name.legend=c("Daphnia","C.brevilabris","P.pediculus","C.floridanus","C.pseudogracilis","C.multisetosum","G.tigrinus","G.zaddachi","Chironomidae","Cricotopus","C.luridus","C.laricomalis","C.sp._BOLD:AAI4299","T.medius","Hydra","H.oligactis","D.polymorpha","P.acuta","S.crystallina","A.chlorotica","B.vejdovskyanum","C.diastrophus","L.hoffmeisteri","P.moldaviensis","S.lacustris","T.tubifex","S.sthenum","A.sieboldi","E.dilatata","L.phaeopa")

col.legend=c("cyan3","cyan3","cyan3","red","red","red","red","red","yellow","yellow","yellow","yellow","yellow","yellow","darkcyan","darkcyan","darkorchid3","darkorchid3","darkorchid3","tan4","tan4","tan4","tan4","tan4","tan4","tan4","grey48","darkgreen","darkgreen","khaki3")
```

After having set up the orders of taxa in top and bottom levels with their corresponding colours, the bipartite plot and the legend can be generated. The legend was generated separately due to limits in the plot dimensions, which if reduced would have made either the plot, or the legend non readable. The legend and the plot were merged externally to R.

```{r plotting the networks}
# Producing the bipartite network
png("R_images/bipartite_network.png", width=4000, height=2500, units="px", res=300)
plotweb(sortweb(food.web, sort.order="seq", sequence=list(seq.higher=seq.high, seq.lower=seq.low)), method = "normal", y.width.low=0.1, y.width.high=0.08,	y.lim=c(0,3), labsize=2, low.lablength=5, low.lab.dis=0.1, high.lab.dis=0.1, low.spacing=0.012, high.spacing = 0.07, text.rot=90, col.high="black", col.low=col.network)
dev.off()

# Producing the legend
png("R_images/bipartite_network_legend.png", width=4000, height=2500, units="px", res=300)
plot.new()
legend("center", xpd=T, legend=name.legend, fill=col.legend, bty="n", border="black")
dev.off()
```

# 6. Diversity analysis of eDNA samples

To avoid over-estimating the diversity present in the eDNA samples, the data set need to be cleaned a bit further. For this purpose assignment at higher taxonomic rank (mainly Order), were removed from the dataset, while assignments to Family, were spearated from the assignments to Genus and Species level. 

```{r eDNA samples rarefaction curves}
clean.edna = clean.edna[,!c(colnames(clean.edna) %in% c("Cyclopoida","Diptera","Podocopida","Spongillida"))]

# splitting Family ranks from Genus and Species ranks
fam.edna = clean.edna[,c(colnames(clean.edna) %in% c("Chironomidae","Chydoridae","Cyclopidae","Daphniidae","Naididae"))]

gen.sp.edna = clean.edna[,!c(colnames(clean.edna) %in% c("Chironomidae","Chydoridae","Cyclopidae","Daphniidae","Naididae"))]

# calculating ratio
gen.sp.edna = gen.sp.edna[]/rowSums(gen.sp.edna)
gen.sp.edna = cbind(gen.sp.edna, info.edna)

# removing `predator_id` variables and transforming eDNA detections into abundances
occupancy.edna = gen.sp.edna[,!c(colnames(gen.sp.edna) %in% c("predator_id"))]
occupancy.edna[,!c(colnames(occupancy.edna) %in% c("time","sample_id","site","sample_type"))][occupancy.edna[,!c(colnames(occupancy.edna) %in% c("time","sample_id","site","sample_type"))] > 0] <- 1

# adding genus information to help in plotting
melt.occupancy.edna = melt(occupancy.edna)
melt.occupancy.edna$genus = gsub("([_]{1}[A-Za-z0-9-]{+}.{+}[.]{0,1}){0,+}$","", melt.occupancy.edna$variable)

# Rarefaction curves calculated on abundance data without genus information
inext.edna = as.data.frame(occupancy.edna[,!c(colnames(occupancy.edna) %in% c("sample_id","sample_type"))] %>% group_by(site, time) %>% summarise_all(funs(sum)))
rownames(inext.edna) = paste(inext.edna$site, inext.edna$time, sep="_")

# removing `time`, `sample_id` and `site` variables as no more necessary.
inext.edna= t(inext.edna[,!c(colnames(inext.edna) %in% c("time","site"))])

# extrapolating rarefaction curves for Species richness and Shannon Diversity
inext.edna = iNEXT(inext.edna, q=c(0,1), datatype = "abundance", se=T, endpoint=200)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")

png("R_images/eDNA_species_rarefactions.png", width = 1000, height = 800, units = "px")
ggiNEXT(inext.edna) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="right") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Species level rarefaction curves of eDNA samples")
dev.off()
```

Calculating the species richness in eDNA samples

```{r species richness in eDNA}
occupancy.edna = as.data.frame(occupancy.edna %>% group_by(site, sample_type, time,sample_id) %>% summarise_all(funs(sum)))
rownames(occupancy.edna) = paste(occupancy.edna$sample_id, occupancy.edna$time, sep="_")

richness.edna = occupancy.edna

richness.edna[grep("eGW|eWB", rownames(richness.edna)), grep("Dikerogammarus", colnames(richness.edna))] <- 0
richness.edna[grep("eRB", rownames(richness.edna)), grep("Gammarus_zaddachi|Gammarus_sp.", colnames(richness.edna))] <- 0

richness.edna = rowSums(occupancy.edna[,!c(colnames(occupancy.edna) %in% c("sample_id","sample_type","site","time"))])

# GW May
gwmay_rich = richness.edna[grep("GW._May", names(richness.edna))]
mean(gwmay_rich)
sd(gwmay_rich)/sqrt(length(gwmay_rich))

# GW Oct
gwoct_rich = richness.edna[grep("GW._Oct", names(richness.edna))]
mean(gwoct_rich)
sd(gwoct_rich)/sqrt(length(gwoct_rich))

# WB May
wbmay_rich = richness.edna[grep("WB._May", names(richness.edna))]
mean(wbmay_rich)
sd(wbmay_rich)/sqrt(length(wbmay_rich))

# WB Oct
wboct_rich = richness.edna[grep("WB._Oct", names(richness.edna))]
mean(wboct_rich)
sd(wboct_rich)/sqrt(length(wboct_rich))

# RB May
rbmay_rich = richness.edna[grep("RB._May", names(richness.edna))]
mean(rbmay_rich)
sd(rbmay_rich)/sqrt(length(rbmay_rich))

# RB Oct
rboct_rich = richness.edna[grep("RB._Oct", names(richness.edna))]
mean(rboct_rich)
sd(rboct_rich)/sqrt(length(rboct_rich))
```

```{r PERMANOVA eDNA}
# eDNA PERMANOVA
rownames(occupancy.edna)=paste(occupancy.edna$sample_id, occupancy.edna$time, sep="_")

adonis.edna = occupancy.edna[,!c(colnames(occupancy.edna) %in% c("time","site","sample_id","sample_type"))]

fact_site.edna = matrix(c(occupancy.edna$site), ncol=1)
fact_time.edna = matrix(c(occupancy.edna$time), ncol=1)

adonis.edna = vegdist(as.matrix(adonis.edna))

adonis(adonis.edna~fact_site.edna*fact_time.edna, permutations = 9999, method="jaccard")
```

```{r nmds based on distances and on abundance data}
# NMDS based on the abundance data (will give the species scores as well)
meta.edna = metaMDS(adonis.edna, distance="jaccard")
```


```{r plotting eDNA nmds}
# full symbols for May, hollow symbols for October
png("R_images/eDNA_NMDS_jaccard.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.edna, type="n", main="eDNA NMDS")#, xlim=c(-1,1.5), ylim=c(-1,1))
points(meta.edna, select=which(fact_site.edna[,1]=="WB" & fact_time.edna[,1]=="May"),col="black",pch=20)
points(meta.edna, select=which(fact_site.edna[,1]=="WB" & fact_time.edna[,1]=="Oct"),col="black",pch=1)
points(meta.edna, select=which(fact_site.edna[,1]=="GW" & fact_time.edna[,1]=="May"),col="blue",pch=15)
points(meta.edna, select=which(fact_site.edna[,1]=="GW" & fact_time.edna[,1]=="Oct"),col="blue",pch=22)
points(meta.edna, select=which(fact_site.edna[,1]=="RB" & fact_time.edna[,1]=="May"),col="red",pch=17)
points(meta.edna, select=which(fact_site.edna[,1]=="RB" & fact_time.edna[,1]=="Oct"),col="red",pch=24)
ordispider(meta.edna, group=fact_site.edna[,1],col=c("blue","orange","green"),label=T)
ordiellipse(meta.edna, group=fact_site.edna[,1],kind="ehull",lty=c(1,3,5),label=F,lwd=1)
text(0,-1, label=paste0("Stress: ", round(meta.edna$stress, 4)))
dev.off()
```

# 7. Community composition from kick-samples.

Microscopy data from the community collected with the kick-samples are available in 6 separate files belonging to each lake from either sampling season.

```{r input files microscopy}
gw.may = read.csv("data/microscopy/gw_may_microscopy.csv", sep="\t")
gw.oct = read.csv("data/microscopy/gw_oct_microscopy.csv", sep="\t")
wb.may = read.csv("data/microscopy/wb_may_microscopy.csv", sep="\t")
wb.oct = read.csv("data/microscopy/wb_oct_microscopy.csv", sep="\t")
rb.may = read.csv("data/microscopy/rb_may_microscopy.csv", sep="\t")
rb.oct = read.csv("data/microscopy/rb_oct_microscopy.csv", sep="\t")
```

Before analysis the data we remove the high rank taxa.

```{r adding the binomial}
data_structuring <- function(df){
  df = as.data.frame(df)
  df = df[,!c(colnames(df) %in% c("ORDER","FAMILY"))]
  df = spread(df, TAXA, NUMBERS, fill=0)
  return(df)
}

gw.may.kick = data_structuring(gw.may)
gw.oct.kick = data_structuring(gw.oct)
wb.may.kick = data_structuring(wb.may)
wb.oct.kick = data_structuring(wb.oct)
rb.may.kick = data_structuring(rb.may)
rb.oct.kick = data_structuring(rb.oct)

# Adding variables

gw.may.kick$time = "May"
gw.oct.kick$time = "Oct"
wb.may.kick$time = "May"
wb.oct.kick$time = "Oct"
rb.may.kick$time = "May"
rb.oct.kick$time = "Oct"

gw.may.kick$sample_type = "kick"
gw.oct.kick$sample_type = "kick"
wb.may.kick$sample_type = "kick"
wb.oct.kick$sample_type = "kick"
rb.may.kick$sample_type = "kick"
rb.oct.kick$sample_type = "kick"

gw.kick = full_join(gw.may.kick, gw.oct.kick)
wb.kick = full_join(wb.may.kick, wb.oct.kick)
rb.kick = full_join(rb.may.kick, rb.oct.kick)

gw.kick[is.na(gw.kick)] <- 0
wb.kick[is.na(wb.kick)] <- 0
rb.kick[is.na(rb.kick)] <- 0
```

```{r}
# Removing unnecessary columns
gw.kick = gw.kick[,!c(colnames(gw.kick) %in% c("TUBE.ID","DATE"))]
wb.kick = wb.kick[,!c(colnames(wb.kick) %in% c("TUBE.ID","DATE"))]
rb.kick = rb.kick[,!c(colnames(rb.kick) %in% c("TUBE.ID","DATE"))]

gw.kick$sample_id = paste0(gw.kick$LOCATION, gw.kick$REPLICATE)
wb.kick$sample_id = paste0(wb.kick$LOCATION, wb.kick$REPLICATE)
rb.kick$sample_id = paste0(rb.kick$LOCATION, rb.kick$REPLICATE)
```

```{r joining frames}
abund.kick = full_join(gw.kick, wb.kick)
abund.kick = full_join(abund.kick, rb.kick)
abund.kick[is.na(abund.kick)] <- 0

colnames(abund.kick)[colnames(abund.kick) == "LOCATION"] <- "site"
```

```{r rarefaction curves on kick-samples}
# removing high order taxa
abund.kick = abund.kick[,!c(colnames(abund.kick) %in% c("Trichoptera","Diptera","Oligochaeta","Ostracoda","Araneae","Coleoptera","Mollusca","Zygoptera","Plecoptera","Nematoda"))]

# setting rownames and removing extra variables
raref.kick = abund.kick[,!c(colnames(abund.kick) %in% c("sample_id","REPLICATE","sample_type"))]

# list of families in data set to be removed
kick.fam = colnames(raref.kick[,grepl("idae", colnames(raref.kick)) == T])
raref.sp.kick = raref.kick[,!c(colnames(raref.kick) %in% kick.fam)]

# summarising data set by lake and season
raref.sp.kick = as.data.frame(raref.sp.kick %>% group_by(site, time) %>% summarise_all(funs(sum)))

rownames(raref.sp.kick) = paste(raref.sp.kick$site, raref.sp.kick$time, sep="_")
raref.sp.kick[is.na(raref.sp.kick)] <- 0
raref.sp.kick = t(raref.sp.kick[,!c(colnames(raref.sp.kick) %in% c("site","time"))])

inext.kick = iNEXT(raref.sp.kick, q=c(0,1), datatype = "abundance", se=T, endpoint=1400)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
png("R_images/kick-samples_rarefactions.png", width = 1000, height = 800, units = "px")
ggiNEXT(inext.kick) +
  labs(y = "", x="Number of individuals") +
  theme(legend.position="right") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Rarefaction curves of communities in kick samples")
dev.off()
```

Calculating PERMANOVA of kick-samples

```{r PERMANOVA of kick-samples}
occupancy.kick = as.data.frame(abund.kick[,!c(colnames(abund.kick) %in% c("REPLICATE"))] %>% group_by(time,site,sample_id,sample_type) %>% summarise_all(funs(sum)))

rownames(occupancy.kick) = paste(occupancy.kick$sample_id, occupancy.kick$time, sep="_")

adonis.kick = occupancy.kick[,!c(colnames(occupancy.kick) %in% c("time","sample_id","site","sample_type"))]
adonis.kick[,!c(colnames(adonis.kick) %in% c("time","sample_id","site","sample_type"))][adonis.kick[,!c(colnames(adonis.kick) %in% c("time","sample_id","site","sample_type"))] > 0] <- 1

fact_site.kick = matrix(c(occupancy.kick$site), ncol=1)
fact_time.kick = matrix(c(occupancy.kick$time), ncol=1)

adonis.kick = vegdist(as.matrix(adonis.kick))

adonis(adonis.kick~fact_site.kick*fact_time.kick, permutations = 9999, method="jaccard")
```

```{r}
meta.kick = metaMDS(adonis.kick, trymax=200, distance="jaccard")
```

```{r NMDS plot of kick-samples}
# full symbols for May, hollow symbols for October
png("R_images/kick-samples_NMDS_jaccard.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.kick, type="n", main="kick-samples NMDS", xlim=c(-0.6,0.6), ylim=c(-0.6,0.6))
points(meta.kick, select=which(fact_site.kick[,1]=="WB" & fact_time.kick[,1]=="May"),col="black",pch=20)
points(meta.kick, select=which(fact_site.kick[,1]=="WB" & fact_time.kick[,1]=="Oct"),col="black",pch=1)
points(meta.kick, select=which(fact_site.kick[,1]=="GW" & fact_time.kick[,1]=="May"),col="blue",pch=15)
points(meta.kick, select=which(fact_site.kick[,1]=="GW" & fact_time.kick[,1]=="Oct"),col="blue",pch=22)
points(meta.kick, select=which(fact_site.kick[,1]=="RB" & fact_time.kick[,1]=="May"),col="red",pch=17)
points(meta.kick, select=which(fact_site.kick[,1]=="RB" & fact_time.kick[,1]=="Oct"),col="red",pch=24)
ordispider(meta.kick, group=fact_site.kick[,1],col=c("blue","orange","green"),label=T)
ordiellipse(meta.kick, group=fact_site.kick[,1],kind="ehull",lty=c(1,3,5),label=F,lwd=1)
text(-0.5,-0.4, label=paste0("Stress: ", round(meta.kick$stress,6)))
dev.off()
```

Calculating species richness for kick-samples

```{r diversity indices kick-samples}
diversity.kick = occupancy.kick
diversity.kick[,!c(colnames(diversity.kick) %in% c("time","sample_id","site","sample_type"))][diversity.kick[,!c(colnames(diversity.kick) %in% c("time","sample_id","site","sample_type"))] > 0] <- 1

diversity.kick[grep("GW|WB", rownames(diversity.kick)), grep("Dikerogammarus", colnames(diversity.kick))] <- 0
diversity.kick[grep("RB", rownames(diversity.kick)), grep("Gammarus_zaddachi|Gammarus_sp.", colnames(diversity.kick))] <- 0

richness.kick = rowSums(diversity.kick[,!c(colnames(diversity.kick) %in% c("sample_id","site","time","sample_type"))])

richness.kick = cbind(richness.kick, occupancy.kick[,c(colnames(occupancy.kick) %in% c("sample_id","sample_type","site","time"))])
colnames(richness.kick)[colnames(richness.kick) == "richness.kick"] <- "richness"
richness.edna = cbind(richness.edna, occupancy.edna[,c(colnames(occupancy.edna) %in% c("sample_id","sample_type","site","time"))])
colnames(richness.edna)[colnames(richness.edna) == "richness.edna"] <- "richness"

richness.full = full_join(richness.edna, richness.kick)
```

```{r richness boxplots}
png("R_images/species_richness_community_boxplot.png", width=3000, height=3000, units="px", res=300)
richness.full %>% mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(site, richness)) +
  geom_boxplot(aes(fill=time), outlier.shape = NA, alpha=0.9) +
  geom_point(position=position_jitterdodge(dodge.width = 0.85), aes(fill=time), color="black", pch=21, size=2, alpha=0.5) +
  facet_grid(.~sample_type, scales = "free_x") +
  theme_bw() + labs(y="Species richness", x="") +
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold"))
dev.off()

shannon.kick = diversity(diversity.kick[,!c(colnames(diversity.kick) %in% c("site","sample_id","sample_type","time"))], index="shannon")
shannon.kick = cbind(shannon.kick, diversity.kick[,c(colnames(diversity.kick) %in% c("site","sample_id","sample_type","time"))])
colnames(shannon.kick)[colnames(shannon.kick) == "shannon.kick"] <- "diversity"

png("R_images/shannon_kick-samples_boxplot.png", width=3000, height=3000, units="px", res=300)
shannon.kick %>% mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(site, diversity)) +
  geom_boxplot(aes(fill=time), outlier.shape = NA, alpha=0.9) +
  geom_point(position=position_jitterdodge(dodge.width = 0.85), aes(fill=time), color="black", pch=21, size=2, alpha=0.5) +
#  facet_grid(.~sample_type, scales = "free_x") +
  theme_bw() + labs(y="Shannon diversity", x="") +
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold"))
dev.off()
```

```{r richness in kick samples}
# GW May
gwmay_kickrich = richness.kick[grep("GW._May", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(gwmay_kickrich)
sd(gwmay_kickrich)/sqrt(length(gwmay_kickrich))

# GW Oct
gwoct_kickrich = richness.kick[grep("GW._Oct", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(gwoct_kickrich)
sd(gwoct_kickrich)/sqrt(length(gwoct_kickrich))

# WB May
wbmay_kickrich = richness.kick[grep("WB._May", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(wbmay_kickrich)
sd(wbmay_kickrich)/sqrt(length(wbmay_kickrich))

# WB Oct
wboct_kickrich = richness.kick[grep("WB._Oct", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(wboct_kickrich)
sd(wboct_kickrich)/sqrt(length(wboct_kickrich))

# RB May
rbmay_kickrich = richness.kick[grep("RB._May", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(rbmay_kickrich)
sd(rbmay_kickrich)/sqrt(length(rbmay_kickrich))

# RB Oct
rboct_kickrich = richness.kick[grep("RB._Oct", rownames(richness.kick)),!c(colnames(richness.kick) %in% c("sample_id","sample_type","site","time"))]
mean(rboct_kickrich)
sd(rboct_kickrich)/sqrt(length(rboct_kickrich))
```


# 8. Community composition of eDNA and kick-samples

```{r merging kick-samples and eDNA frames}
occup.kick.edna = full_join(occupancy.edna, occupancy.kick)
occup.kick.edna[is.na(occup.kick.edna)] <- 0

adonis.kick.edna = occup.kick.edna[,!c(colnames(occup.kick.edna) %in% c("site","sample_id","sample_type","time","REPLICATE"))]

fact_site.edki = matrix(c(occup.kick.edna$site), ncol=1)
fact_time.edki = matrix(c(occup.kick.edna$time), ncol=1)
fact_type.edki = matrix(c(occup.kick.edna$sample_type), ncol=1)

adonis(adonis.kick.edna~fact_site.edki*fact_time.edki*fact_type.edki, permutations = 9999, method = "jaccard")
```

```{r}
meta.kick.edna = metaMDS(adonis.kick.edna, trymax = 200, methods = "jaccard")
```


```{r NMDS plot of kick-samples and eDNA}
# full symbols for May, hollow symbols for October

for (s in grep("_", rownames(meta.kick.edna$species))){
  if (grepl("_sp.|cf.", rownames(meta.kick.edna$species)[s]) == T){
    rownames(meta.kick.edna$species)[s] = rownames(meta.kick.edna$species)[s]
    } else {
      rownames(meta.kick.edna$species)[s] = gsub("[a-z]{3,13}_",".", rownames(meta.kick.edna$species)[s])
    }
}

png("R_images/kick-eDNA_NMDS_jaccard.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.kick.edna, type="n", main="Community NMDS\neDNA and kick-samples by sample type", xlim=c(-1.5,1.5), ylim=c(-1,1.1))
points(meta.kick.edna, select=which(fact_site.edki[,1]=="WB" & fact_time.edki[,1]=="May"),col="black",pch=20)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="WB" & fact_time.edki[,1]=="Oct"),col="black",pch=1)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="GW" & fact_time.edki[,1]=="May"),col="blue",pch=15)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="GW" & fact_time.edki[,1]=="Oct"),col="blue",pch=22)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="RB" & fact_time.edki[,1]=="May"),col="red",pch=17)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="RB" & fact_time.edki[,1]=="Oct"),col="red",pch=24)
orditorp(meta.kick.edna, display="species")
ordispider(meta.kick.edna, group=fact_site.edki[,1],col=c("blue","orange","green"),label=T)
ordiellipse(meta.kick.edna, group=fact_type.edki[,1],kind="ehull",lty=c(1,3),label=F,lwd=1)
text(0, -1, label=paste0("Stress: ",round(meta.kick.edna$stress, 6)))
dev.off()
```

```{r envfit of communities from October}
# loading the file
water.measure = read.csv("data/env_measures/water_measure.csv", sep=",", header=T, row.names = 1)
occup.oct = occup.kick.edna[occup.kick.edna$time == "Oct",]

env.measures = as.data.frame(water.measure %>% transmute(temp_mean = (temp1+temp2+temp3)/3,
                            pH_mean = (pH1+pH2+pH3)/3,
                            cond_mean = (cond1.uS.+cond2.uS.+cond3.uS.)/3))

rownames(env.measures) = rownames(water.measure)
rownames(occup.oct) = occup.oct$sample_id
env.measures$sample_id = rownames(water.measure)

env.measures = left_join(env.measures, occup.oct[grep("e", rownames(occup.oct), invert=T),c("Dikerogammarus_villosus", "Dikerogammarus_sp.","sample_id")])

env.measures$Dikerogammarus_villosus = env.measures$Dikerogammarus_sp. + env.measures$Dikerogammarus_villosus

occup.oct = occup.oct[,!c(colnames(occup.oct) %in% c("time","sample_id","site","sample_type","Dikerogammarus_sp.","Dikerogammaus_villosus"))]
env.measures = env.measures[,!c(colnames(env.measures) %in% c("sample_id", "Dikerogammarus_sp."))]

envfit.comm = as.data.frame(t(occup.oct))

sum.comm = as.data.frame(envfit.comm %>% transmute(GW1 = GW1+eGW1,
                        GW2 = GW2+eGW2,
                        GW3 = GW3+eGW3,
                        GW4 = GW4+eGW4,
                        GW5 = eGW5,
                        GW6 = GW6+eGW6,
                        WB1 = WB1+eWB1,
                        WB2 = WB2+eWB2,
                        WB3 = WB3+eWB3,
                        WB4 = WB4+eWB4,
                        WB5 = WB5+eWB5,
                        WB6 = WB6+eWB6,
                        RB1 = RB1+eRB1,
                        RB2 = RB2+eRB2,
                        RB3 = RB3,
                        RB4 = RB4+eRB4,
                        RB5 = RB5+eRB5,
                        RB6 = RB6+eRB6))

sum.comm = as.data.frame(t(sum.comm))
colnames(sum.comm) = colnames(occup.oct)

ord = metaMDS(sum.comm, distance = "jaccard", trymax = 200)
fit.oct = envfit(ord, env.measures, permu = 9999, na.rm=T, labels=T)
```

```{r envfit plot}
png("R_images/envift_on_October_communities.png", width = 3000, height = 3000, units = "px", res = 300)
plot(ord, type="n", xlim=c(-1.5,1.3), ylim=c(-1.3,1.1), cex.axis=2, cex.lab=1.5)
abline(h=0, v=0, lty=3, lwd=2, col="grey50")
orditorp(ord, display="sites", cex=1, col=c("blue","blue","blue","blue","blue","blue","darkgreen","darkgreen","darkgreen","darkgreen","darkgreen","darkgreen","darkorange","darkorange","darkorange","darkorange","darkorange","darkorange"))
orditorp(ord, display="species", type="n")
plot(fit.oct, col="red", cex=2)
dev.off()
```

# 9. Comparison of gut contents detection against community composition

```{r merging guts, and community data sets}
# Previously the sample type for the gut contents data frame was removed, so we first add it again.
abund.guts.site$sample_type = "guts"
occup.kick.edna$sample_type

occup.gut.edna.kick = full_join(abund.guts.site, occup.kick.edna)
occup.gut.edna.kick[is.na(occup.gut.edna.kick)] <- 0

occup.gut.edna.kick = as.data.frame(occup.gut.edna.kick[,!c(colnames(occup.gut.edna.kick) %in% c("predator_id"))] %>% group_by(sample_type,site,time,sample_id) %>% summarise_all(funs(sum)))

adonis.guts.edna.kick = occup.gut.edna.kick[,!c(colnames(occup.gut.edna.kick) %in% c("site","sample_type","time","predator_id","sample_id"))]

fact_site.full = matrix(c(occup.gut.edna.kick$site), ncol=1)
fact_time.full = matrix(c(occup.gut.edna.kick$time), ncol=1)
fact_type.full = matrix(c(occup.gut.edna.kick$sample_type), ncol=1)

adonis.guts.edna.kick = vegdist(as.matrix(adonis.guts.edna.kick))

adonis(adonis.guts.edna.kick~fact_type.full*fact_site.full*fact_time.full, permutations = 9999, method = "jaccard")
```

```{r permanova of guts, eDNA and kick-samples}
meta.full = metaMDS(adonis.guts.edna.kick, trymax = 200, methods = "jaccard")
```


```{r NMDS plot of guts against kick-samples and eDNA}
# full symbols for May, hollow symbols for October
png("R_images/guts_community_NMDS_jaccard.png", width = 3000, height = 3000, units = "px", res = 300)
plot(meta.full, type="n", main="Gut contents and Community NMDS", xlim=c(-0.5,0.5), ylim=c(-0.6,0.7))
points(meta.full, select=which(fact_site.full[,1]=="WB" & fact_time.full[,1]=="May"),col="black",pch=20)
points(meta.full, select=which(fact_site.full[,1]=="WB" & fact_time.full[,1]=="Oct"),col="black",pch=1)
points(meta.full, select=which(fact_site.full[,1]=="GW" & fact_time.full[,1]=="May"),col="blue",pch=15)
points(meta.full, select=which(fact_site.full[,1]=="GW" & fact_time.full[,1]=="Oct"),col="blue",pch=22)
points(meta.full, select=which(fact_site.full[,1]=="RB" & fact_time.full[,1]=="May"),col="red",pch=17)
points(meta.full, select=which(fact_site.full[,1]=="RB" & fact_time.full[,1]=="Oct"),col="red",pch=24)
ordispider(meta.full, group=fact_site.full[,1],col=c("blue","orange","green"),label=T)
ordiellipse(meta.full, group=fact_type.full[,1],kind="ehull",lty=c(1,5,3),label=F,lwd=1)
text(0, -0.5, label=paste0("Stress: ",round(meta.full$stress, 6)))
dev.off()
```


```{r barplot of guts and community}
occup.melt = melt(occup.gut.edna.kick)

# Adding order groups to the data frame
occup.melt$order = "NA"
occup.melt[grep("Limnesia|Platynothrus", occup.melt$variable),"order"] = "Arachnida"
occup.melt[grep("Chydoridae|Chydorus|Daphnia|Daphniidae|Diaphanosoma|Eubosmina|Eurycercus|Macrothrix|Polyphemus", occup.melt$variable),"order"] = "Cladocera"
occup.melt[grep("Dytiscidae|Elmidae|Haliplidae|Limnius|Platambus", occup.melt$variable), "order"] = "Coleoptera"
occup.melt[grep("Ploima|Xenylla|Podura", occup.melt$variable), "order"] = "Collembola"
occup.melt[grep("Asellus", occup.melt$variable), "order"] = "Isopoda"
occup.melt[grep("Corophium|Crangonyx|Dikerogammarus|Gammaridae|Gammarus", occup.melt$variable), "order"] = "Amphipoda"
occup.melt[grep("Acanthocyclops|Cyclops|Eucyclops|Eudiaptomus|Eurytemora|Mesocyclops", occup.melt$variable), "order"] = "Cyclopoida"
occup.melt[grep("Polypedilum|Brachycera|Ceratopogonidae|Chironomidae|Chironomus|Cricotopus|Culicoides|Dolichopodidae|Nematocera|Polypedilum|Procladius|Tanytarsus|Psychodidae|Tipulidae", occup.melt$variable), "order"]= "Diptera"
occup.melt[grep("Caenis|Cloeon", occup.melt$variable), "order"] = "Ephemeroptera"
occup.melt[grep("Chaetonotus", occup.melt$variable), "order"] = "Gastrotricha"
occup.melt[grep("Corixa|Corixidae|Micronecta|Nepa|Notonecta|Plea|Velliidae", occup.melt$variable), "order"] = "Hemiptera"
occup.melt[grep("Placobdella|Erpobdella|Glossiphonia|Helobdella", occup.melt$variable), "order"] = "Hirudinea"
occup.melt[grep("Hydra", occup.melt$variable), "order"] = "Hydrozoa"
occup.melt[grep("Sialidae", occup.melt$variable), "order"] = "Megaloptera"
occup.melt[grep("Theodoxus|Stagnicola|Valvatidae|Segmentina|Anodonta|Bithynia|Bithyniidae|Corbicula|Dreissena|Hydrobiidae|Lymnaea|Lymnaeidae|Physa|Physella|Pisidium|Planorbidae|Planorbis|Radix|Sida|Sphaerium|Unio|Valvata|Potamopyrgus", occup.melt$variable), "order"] = "Mollusca"
occup.melt[grep("Plectus", occup.melt$variable), "order"] = "Nematoda"
occup.melt[grep("Bothrioneurum|Allolobophora_chlorotica|Chaetogaster|Lumbricillus|Nais|Ophidonais|Polyarthra|Potamothrix|Stylaria|Synchaeta|Tubifex|Tubificidae|Limnodrilus|Uncinais|Cyprideis|Cypridopsis", occup.melt$variable), "order"] = "Oligochaeta"
occup.melt[grep("Limnocythere", occup.melt$variable), "order"] = "Ostracoda"
occup.melt[grep("Stenostomum|Stempellinella", occup.melt$variable), "order"] = "Platyhelminthes"
occup.melt[grep("Polyarthra|Asplanchna|Brachionus|Euchlanis|Keratella|Philodina", occup.melt$variable), "order"] = "Rotifera"
occup.melt[grep("Ceraclea|Ecnomus|Leptoceridae|Lype|Oecetis|Oxyethira", occup.melt$variable), "order"] = "Trichoptera"
occup.melt[grep("Coenagrionidae|Erythromma", occup.melt$variable), "order"] = "Zygoptera"
occup.melt[grep("Liposcelis", occup.melt$variable), "order"] = "Psocoptera"

cols_barplot = c("red","darkorange", "cyan3", "greenyellow", "plum2", "deeppink", "yellow", "rosybrown",  "green","salmon","palevioletred","aquamarine","goldenrod","khaki4","darkorchid3","gainsboro","tan4", "brown","grey50","black","darkgreen","khaki3","blue")

png("R_images/barplot_summary_by_order.png", width = 3000, height = 3000, units = "px", res = 300)
occup.melt %>% mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(sample_type, value)) +
  geom_bar(stat="identity", position="fill", aes(fill=order)) +
  scale_fill_manual(values = cols_barplot, name = "") +
  scale_x_discrete(limits=c("eDNA","kick","guts")) +
  theme_bw() +
  facet_grid(time~site) +
  labs(x="", y="Community composition (%)", title = "Community composition (individual taxa grouped by high rank taxonomy)") +
  theme(legend.position = "right")
dev.off()
```

```{r summary barplot of amphipod}
melt.kick = melt(abund.kick[,!c(colnames(abund.kick) %in% "REPLICATE")])

melt.kick[grep("Dikerogammarus", melt.kick$variable),c("sample_id","time","value")]

melt.kick$order = "NA"
melt.kick[grep("Limnesia|Platynothrus", melt.kick$variable),"order"] = "Arachnida"
melt.kick[grep("Chydoridae|Chydorus|Daphnia|Daphniidae|Diaphanosoma|Eubosmina|Eurycercus|Macrothrix|Polyphemus", melt.kick$variable),"order"] = "Cladocera"
melt.kick[grep("Dytiscidae|Elmidae|Haliplidae|Limnius|Platambus", melt.kick$variable), "order"] = "Coleoptera"
melt.kick[grep("Ploima|Xenylla|Podura", melt.kick$variable), "order"] = "Collembola"
melt.kick[grep("Asellus", melt.kick$variable), "order"] = "Isopoda"
melt.kick[grep("Corophium|Crangonyx|Dikerogammarus|Gammarus", melt.kick$variable), "order"] = "Amphipoda"
melt.kick[grep("Acanthocyclops|Cyclops|Eucyclops|Eudiaptomus|Eurytemora|Mesocyclops", melt.kick$variable), "order"] = "Cyclopoida"
melt.kick[grep("Polypedilum|Brachycera|Ceratopogonidae|Chironomidae|Chironomus|Cricotopus|Culicoides|Dolichopodidae|Nematocera|Polypedilum|Procladius|Tanytarsus|Psychodidae|Tipulidae", melt.kick$variable), "order"]= "Diptera"
melt.kick[grep("Caenis|Cloeon", melt.kick$variable), "order"] = "Ephemeroptera"
melt.kick[grep("Chaetonotus", melt.kick$variable), "order"] = "Gastrotricha"
melt.kick[grep("Corixa|Corixidae|Micronecta|Nepa|Notonecta|Plea|Velliidae", melt.kick$variable), "order"] = "Hemiptera"
melt.kick[grep("Placobdella|Erpobdella|Glossiphonia|Helobdella", melt.kick$variable), "order"] = "Hirudinea"
melt.kick[grep("Hydra", melt.kick$variable), "order"] = "Hydrozoa"
melt.kick[grep("Sialidae", melt.kick$variable), "order"] = "Megaloptera"
melt.kick[grep("Theodoxus|Stagnicola|Valvatidae|Segmentina|Anodonta|Bithynia|Bithyniidae|Corbicula|Dreissena|Hydrobiidae|Lymnaea|Lymnaeidae|Physa|Physella|Pisidium|Planorbidae|Planorbis|Radix|Sida|Sphaerium|Unio|Valvata|Potamopyrgus", melt.kick$variable), "order"] = "Mollusca"
melt.kick[grep("Plectus", melt.kick$variable), "order"] = "Nematoda"
melt.kick[grep("Bothrioneurum|Allolobophora_chlorotica|Chaetogaster|Lumbricillus|Nais|Ophidonais|Polyarthra|Potamothrix|Stylaria|Synchaeta|Tubifex|Tubificidae|Limnodrilus|Uncinais|Cyprideis|Cypridopsis", melt.kick$variable), "order"] = "Oligochaeta"
melt.kick[grep("Limnocythere", melt.kick$variable), "order"] = "Ostracoda"
melt.kick[grep("Stenostomum|Stempellinella", melt.kick$variable), "order"] = "Platyhelminthes"
melt.kick[grep("Polyarthra|Asplanchna|Brachionus|Euchlanis|Keratella|Philodina", melt.kick$variable), "order"] = "Rotifera"
melt.kick[grep("Ceraclea|Ecnomus|Leptoceridae|Lype|Oecetis|Oxyethira", melt.kick$variable), "order"] = "Trichoptera"
melt.kick[grep("Coenagrionidae|Erythromma", melt.kick$variable), "order"] = "Zygoptera"
melt.kick[grep("Liposcelis", melt.kick$variable), "order"] = "Psocoptera"

melt.kick$variable = gsub("_"," ", melt.kick$variable)

col_crustacea = c("olivedrab","springgreen","darkblue","steelblue1","skyblue1","darkorchid2","plum","orchid1","tomato","tan1")
#c("#00CC99","#006666","#3399CC","#0000FF","#000033","#9900FF","#660066","#CC00CC","#FF00FF","#9966FF","#808080")

png("R_images/barplot_summary_amphipod.png", width = 3000, height = 3000, units = "px", res = 300)
melt.kick[melt.kick$order == "Amphipoda",] %>% 
  mutate(variable = factor(variable, levels=c("Corophium sp.", "Corophium curvispinum", "Crangonyx sp.", "Crangonyx floridanus", "Crangonyx pseudogracilis", "Gammarus sp.", "Gammarus tigrinus", "Gammarus zaddachi", "Dikerogammarus sp.", "Dikerogammarus villosus"))) %>% 
  mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(sample_id, value, group=time)) +
  geom_bar(stat="identity", aes(fill=variable)) +
  scale_fill_manual(values = col_crustacea, name = "") +
  theme_bw() +
  facet_grid(time~site, scales="free_x") +
  labs(x="", y="Abundance (count)", title = "Amphipoda kick-samples composition") +
  theme(legend.position = "right")
dev.off()
```

# 10. Generating data summary table

This last chunk is only to create the base for the summary table presented in supplementary informaton. This uses the presence/absence data for the gut contents and the eDNA samples, while it uses the actual abundance for the kick-samples. The table was then expanded externally to R to include more summarising informations.

```{r generating data summary table}
guts.summary = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("sample_id"))] %>% group_by(site,time,predator_id,sample_type) %>% summarise_all(funs(sum)))
edna.summary = as.data.frame(occupancy.edna[,!c(colnames(occupancy.edna) %in% c("sample_id"))] %>% group_by(site,time, sample_type) %>% summarise_all(funs(sum)))
kick.summary = as.data.frame(abund.kick[,!c(colnames(abund.kick) %in% c("sample_id","REPLICATE"))] %>% group_by(site,time, sample_type) %>% summarise_all(funs(sum)))

full.data = full_join(guts.summary, edna.summary)
full.data = full_join(full.data, kick.summary)

# Splitting the metadata from data, so that we can remove the NAs into the data set.
full.info = full.data[,c(colnames(full.data) %in% c("sample_type","predator_id","site","time"))]
full.data = full.data[,!c(colnames(full.data) %in% c("sample_type","predator_id","site","time"))]
full.data[is.na(full.data)] <- 0

# Binding the frame together using specific order of the metadata
final.data = bind_cols(full.info[,c(4,3,1,2)], full.data)

# Transposing and saving the data frame, we will rearrange the order of the metadata later
final.data = t(final.data)

#Saving final output
write.csv(final.data, "R_outputs/Data_summary_table.csv")
```

# 11. Producing plots for publication

```{r publication quality NMDS plot}
png("R_images/NMDS_multiplot.png",width = 4000, height = 4000, units = "px", res = 310)
lay.matrix = matrix(c(1,2,3,4,5,5), ncol=2, nrow=3, byrow=T)
layout(lay.matrix, heights=c(2,2,1), widths=c(2,2))

# NMDS eDNA only
plot(meta.edna, type="n", xlim=c(-0.4,0.5), ylim=c(-0.5,0.4), cex.axis=2, cex.lab=1.5)
points(meta.edna, select=which(fact_site.edna[,1]=="WB" & fact_time.edna[,1]=="May"),col="black",pch=20, cex=2)
points(meta.edna, select=which(fact_site.edna[,1]=="WB" & fact_time.edna[,1]=="Oct"),col="black",pch=1, cex=2)
points(meta.edna, select=which(fact_site.edna[,1]=="GW" & fact_time.edna[,1]=="May"),col="blue",pch=15, cex=2)
points(meta.edna, select=which(fact_site.edna[,1]=="GW" & fact_time.edna[,1]=="Oct"),col="blue",pch=22, cex=2)
points(meta.edna, select=which(fact_site.edna[,1]=="RB" & fact_time.edna[,1]=="May"),col="red",pch=17, cex=2)
points(meta.edna, select=which(fact_site.edna[,1]=="RB" & fact_time.edna[,1]=="Oct"),col="red",pch=24, cex=2)
ordispider(meta.edna, group=fact_site.edna[,1],col=c("blue","orange","green"),label=T, cex=2)
ordiellipse(meta.edna, group=fact_site.edna[,1],kind="ehull",lty=c(1,3,5),label=F,lwd=1)
text(0.4,-0.4, label=paste0("Stress: ", round(meta.edna$stress, 4)), cex=1.5)
text(-0.55,0.35, label="a)", cex=3)
text(-0.5,-0.4,label="eDNA", cex=2.5)

# NMDS kick-samples only
plot(meta.kick, type="n", xlim=c(-0.5,0.5), ylim=c(-0.6,0.6), cex.axis=2, cex.lab=1.5)
points(meta.kick, select=which(fact_site.kick[,1]=="WB" & fact_time.kick[,1]=="May"),col="black",pch=20, cex=2)
points(meta.kick, select=which(fact_site.kick[,1]=="WB" & fact_time.kick[,1]=="Oct"),col="black",pch=1, cex=2)
points(meta.kick, select=which(fact_site.kick[,1]=="GW" & fact_time.kick[,1]=="May"),col="blue",pch=15, cex=2)
points(meta.kick, select=which(fact_site.kick[,1]=="GW" & fact_time.kick[,1]=="Oct"),col="blue",pch=22, cex=2)
points(meta.kick, select=which(fact_site.kick[,1]=="RB" & fact_time.kick[,1]=="May"),col="red",pch=17, cex=2)
points(meta.kick, select=which(fact_site.kick[,1]=="RB" & fact_time.kick[,1]=="Oct"),col="red",pch=24, cex=2)
ordispider(meta.kick, group=fact_site.kick[,1],col=c("blue","orange","green"),label=T, cex=2)
ordiellipse(meta.kick, group=fact_site.kick[,1],kind="ehull",lty=c(1,3,5),label=F,lwd=1)
text(0.6,-0.5, label=paste0("Stress: ", round(meta.kick$stress, 4)), cex=1.5)
text(-0.8,0.53, label="b)", cex=3)
text(-0.7,-0.5,label="kick", cex=2.5)

# NMDS eDNA vs kick-samples
plot(meta.kick.edna, type="n", xlim=c(-0.8,0.8), ylim=c(-0.7,0.7), cex.axis=2, cex.lab=1.5)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="WB" & fact_time.edki[,1]=="May"),col="black",pch=20, cex=2)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="WB" & fact_time.edki[,1]=="Oct"),col="black",pch=1, cex=2)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="GW" & fact_time.edki[,1]=="May"),col="blue",pch=15, cex=2)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="GW" & fact_time.edki[,1]=="Oct"),col="blue",pch=22, cex=2)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="RB" & fact_time.edki[,1]=="May"),col="red",pch=17, cex=2)
points(meta.kick.edna, select=which(fact_site.edki[,1]=="RB" & fact_time.edki[,1]=="Oct"),col="red",pch=24, cex=2)
orditorp(meta.kick.edna, display="species", type="n", cex=1.2)
ordispider(meta.kick.edna, group=fact_site.edki[,1],col=c("blue","orange","green"),label=T, cex=2)
ordiellipse(meta.kick.edna, group=fact_type.edki[,1],kind="ehull",lty=c(1,3),label=T,lwd=1)
text(0.7,-0.6, label=paste0("Stress: ",round(meta.kick.edna$stress, 4)), cex=1.5)
text(-0.65,0.7, label="eDNA", cex=2.5)
text(0.75,0.57, label="kick", cex=2.5)
text(-0.95,0.63, label="c)", cex=3)

# envfit of community
plot(ord, type="n", xlim=c(-1.5,1.3), ylim=c(-1.3,1.1), cex.axis=2, cex.lab=1.5)
abline(h=0, v=0, lty=3, lwd=2, col="grey50")
orditorp(ord, display="sites", col=c("blue","blue","blue","blue","blue","blue","darkgreen","darkgreen","darkgreen","darkgreen","darkgreen","darkgreen","darkorange","darkorange","darkorange","darkorange","darkorange","darkorange"), cex=2)
orditorp(ord, display="species", type="n", cex=1.2)
plot(fit.oct, col="red", cex=2, p.max=0.1)
text(-1.7,1, label="d)", cex=3)
text(-1.5,-1.2,label="envfit", cex=2.5)

# Common legend
plot.new()
legend("top", legend=c("Grafham Water - May","Grafham Water - Oct","Wroxham Broad - May","Wroxham Broad - Oct","Rockand Broad - May","Rockand Broad - Oct"), col=c("blue","blue","black","black","red","red"), pch=c(15,22,19,1,17,24), cex=2, ncol=3)
dev.off()
```


```{r publication quality Dv-Gz multiplot}
png("R_images/Dv-Gz_multiplot.png",width = 4000, height = 4000, units = "px", res = 300)
lay.matrix = matrix(c(1,2,3,3), ncol=2, nrow=2, byrow=T)
layout(lay.matrix, heights=c(2,0.5), widths=c(2,2))

# NMDS of gut contents Dv-Gz
plot(meta.guts, type="n", xlim=c(-1.5,1.5), ylim=c(-1.2,1.5), cex.axis=2, cex.lab=1.5)
points(meta.guts, select=which(fact_site.guts[,1]=="WB" & fact_time.guts[,1]=="May"),col="black",pch=20, cex=2)
points(meta.guts, select=which(fact_site.guts[,1]=="WB" & fact_time.guts[,1]=="Oct"),col="black",pch=1, cex=2)
points(meta.guts, select=which(fact_site.guts[,1]=="GW" & fact_time.guts[,1]=="May"),col="blue",pch=15, cex=2)
points(meta.guts, select=which(fact_site.guts[,1]=="GW" & fact_time.guts[,1]=="Oct"),col="blue",pch=22, cex=2)
points(meta.guts, select=which(fact_site.guts[,1]=="RB" & fact_time.guts[,1]=="May"),col="red",pch=17, cex=2)
points(meta.guts, select=which(fact_site.guts[,1]=="RB" & fact_time.guts[,1]=="Oct"),col="red",pch=24, cex=2)
ordiellipse(meta.guts,group=fact_predator.guts[,1],kind="ehull",lty=c(1,5),label=F,lwd=1)
ordispider(meta.guts,group=fact_predator.guts[,1],col=c("blue","orange"),label=T, cex=2)
text(1,-1.6,label=paste0("Stress: ",round(meta.guts$stress, 4)), cex=1.5)
text(-1.4, 2.5, label="a)", cex=3)
text(-1.1, 0.8, label="D.villosus", cex=2.5)
text(1, 1.1, label="G.zaddachi", cex=2.5)

# NMDS of guts against eDNA and kick samples
plot(meta.full, type="n", xlim=c(-0.5,0.5), ylim=c(-0.6,0.7), cex.axis=2, cex.lab=1.5)
points(meta.full, select=which(fact_site.full[,1]=="WB" & fact_time.full[,1]=="May"),col="black",pch=20, cex=2)
points(meta.full, select=which(fact_site.full[,1]=="WB" & fact_time.full[,1]=="Oct"),col="black",pch=1, cex=2)
points(meta.full, select=which(fact_site.full[,1]=="GW" & fact_time.full[,1]=="May"),col="blue",pch=15, cex=2)
points(meta.full, select=which(fact_site.full[,1]=="GW" & fact_time.full[,1]=="Oct"),col="blue",pch=22, cex=2)
points(meta.full, select=which(fact_site.full[,1]=="RB" & fact_time.full[,1]=="May"),col="red",pch=17, cex=2)
points(meta.full, select=which(fact_site.full[,1]=="RB" & fact_time.full[,1]=="Oct"),col="red",pch=24, cex=2)
ordispider(meta.full, group=fact_site.full[,1],col=c("blue","orange","green"),label=T, cex=2)
ordiellipse(meta.full, group=fact_type.full[,1],kind="ehull",lty=c(1,5,3),label=F,lwd=1)
text(0.3, -0.55, label=paste0("Stress: ",round(meta.full$stress, 4)), cex=1.5)
text(-0.45, 0.83, label="b)", cex=3)
text(-0.4, 0.4, label="eDNA", cex=2.5)
text(0, 0.7, label="gut", cex=2.5)
text(0.4, 0.4, label="kick", cex=2.5)

# NMDS legend
plot.new()
legend("top", legend=c("Grafham Water - May","Grafham Water - Oct","Wroxham Broad - May","Wroxham Broad - Oct","Rockand Broad - May","Rockand Broad - Oct"), col=c("blue","blue","black","black","red","red"), pch=c(15,22,19,1,17,24), cex=2, ncol=3)
dev.off()
```

```{r bipartite publication}
png("R_images/Bipartite_publication.png",width = 6000, height = 4000, units = "px", res = 300)
# legend for multiplot
name.multi=c("Cladocera","Amphipoda","Diptera","Hydrozoa","Mollusca","Oligochaeta","Platyhelminthes","Rotifera","Trichoptera")
col.multi=c("cyan3","red","yellow","darkcyan","darkorchid3","tan4","grey48","darkgreen","khaki3")

# Bipartite network
plotweb(sortweb(food.web, sort.order="seq", sequence=list(seq.higher=seq.high, seq.lower=seq.low)), method = "normal", y.width.low=0.1, y.width.high=0.08,	y.lim=c(0,3), labsize=2, low.lab.dis=0.1, high.lab.dis=0.1, low.spacing=0.012, high.spacing = 0.07, text.rot=90, col.high="black", col.low=col.network, high.y=2, low.y=1)
legend("bottom", legend=name.multi, fill=col.multi, border="black", bty="n", ncol=3, xpd=T, cex=1.7)
dev.off()
```

```{r multiplot for the summary barplots}
png("R_images/Summary_barplot_multiplot.png", width = 4000, height = 5000, units = "px", res = 300)
# barplot of communities
a = occup.melt %>% mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(sample_type, value)) +
  geom_bar(stat="identity", position="fill", aes(fill=order)) +
  scale_fill_manual(values = cols_barplot, name = "") +
  scale_x_discrete(limits=c("eDNA","kick","guts")) +
  theme_bw() +
  facet_grid(time~site) +
  labs(x="", y="Composition (%)") +
  guides(fill=guide_legend(ncol=1)) +
  theme(legend.position = "right", axis.text = element_text(size = 15), 
        strip.text = element_text(size=15), axis.title = element_text(size =15),
        legend.text = element_text(size=15))

# barplot of amphipod
b = melt.kick[melt.kick$order == "Amphipoda",] %>% 
  mutate(variable = factor(variable, levels=c("Corophium sp.", "Corophium curvispinum", "Crangonyx sp.", "Crangonyx floridanus", "Crangonyx pseudogracilis", "Gammarus sp.", "Gammarus tigrinus", "Gammarus zaddachi", "Dikerogammarus sp.", "Dikerogammarus villosus"))) %>% 
  mutate(site = factor(site, levels=c("GW","WB","RB"))) %>% 
  ggplot(aes(sample_id, value, group=time)) +
  geom_bar(stat="identity", aes(fill=variable)) +
  scale_fill_manual(values = col_crustacea, name = "") +
  theme_bw() +
  facet_grid(time~site, scales="free_x") +
  labs(x="", y="Abundance (count)") +
  theme(legend.position = "right", axis.text = element_text(size = 15), 
        strip.text = element_text(size=15), axis.title = element_text(size =15),
        legend.text = element_text(size=15))

plot_grid(a, b, labels=c("a)","b)"), nrow=2, label_size=25, label_fontface=1)

dev.off()
```

```{r log session info}
sink("log_session_analysis.txt")
sessionInfo()
citation()
sink()
```

```{r packages citation}
sink("packages_citations.txt")
for (p in pack.list){
 print(citation(p))
}
sink()
```

```{r}
print("end")
```
